<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Chen Yang" />

<meta name="date" content="2026-02-07" />

<title>Advanced Features &amp; Case Studies</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Advanced Features &amp; Case Studies</h1>
<h4 class="author">Chen Yang</h4>
<h4 class="date">2026-02-07</h4>



<p>This article explores advanced features of mLLMCelltype and presents
practical examples demonstrating its application in various research
contexts.</p>
<div id="hierarchical-cell-type-annotation" class="section level2">
<h2>Hierarchical Cell Type Annotation</h2>
<div id="understanding-hierarchical-annotation" class="section level3">
<h3>Understanding Hierarchical Annotation</h3>
<p>Cell types often exist in hierarchical relationships. For example, T
cells can be further classified into CD4+ T cells, CD8+ T cells,
regulatory T cells, etc. mLLMCelltype can be used in a multi-step
workflow to capture these hierarchical relationships.</p>
</div>
<div id="implementing-hierarchical-annotation" class="section level3">
<h3>Implementing Hierarchical Annotation</h3>
<p>Here’s a practical approach to perform hierarchical annotation:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(mLLMCelltype)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="co"># Step 1: Perform initial high-level annotation</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a>high_level_results <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a>  <span class="at">input =</span> marker_data,</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,</span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a>  <span class="at">api_key =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>),</span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a>  <span class="at">top_gene_count =</span> <span class="dv">10</span></span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a>)</span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a><span class="co"># Step 2: Add high-level annotations to Seurat object</span></span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a>seurat_obj<span class="sc">$</span>high_level_celltype <span class="ot">&lt;-</span> plyr<span class="sc">::</span><span class="fu">mapvalues</span>(</span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">as.character</span>(<span class="fu">Idents</span>(seurat_obj)),</span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a>  <span class="at">from =</span> <span class="fu">names</span>(high_level_results),</span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a>  <span class="at">to =</span> high_level_results</span>
<span id="cb1-19"><a href="#cb1-19" tabindex="-1"></a>)</span>
<span id="cb1-20"><a href="#cb1-20" tabindex="-1"></a></span>
<span id="cb1-21"><a href="#cb1-21" tabindex="-1"></a><span class="co"># Step 3: Subset T cells for further annotation</span></span>
<span id="cb1-22"><a href="#cb1-22" tabindex="-1"></a>t_cells <span class="ot">&lt;-</span> <span class="fu">subset</span>(seurat_obj, high_level_celltype <span class="sc">==</span> <span class="st">&quot;T cells&quot;</span>)</span>
<span id="cb1-23"><a href="#cb1-23" tabindex="-1"></a></span>
<span id="cb1-24"><a href="#cb1-24" tabindex="-1"></a><span class="co"># Step 4: Find markers within T cells</span></span>
<span id="cb1-25"><a href="#cb1-25" tabindex="-1"></a>t_cell_markers <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(t_cells, <span class="at">only.pos =</span> <span class="cn">TRUE</span>, <span class="at">min.pct =</span> <span class="fl">0.25</span>, <span class="at">logfc.threshold =</span> <span class="fl">0.25</span>)</span>
<span id="cb1-26"><a href="#cb1-26" tabindex="-1"></a></span>
<span id="cb1-27"><a href="#cb1-27" tabindex="-1"></a><span class="co"># Step 5: Perform T cell subtype annotation</span></span>
<span id="cb1-28"><a href="#cb1-28" tabindex="-1"></a>t_cell_subtypes <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb1-29"><a href="#cb1-29" tabindex="-1"></a>  <span class="at">input =</span> t_cell_markers,</span>
<span id="cb1-30"><a href="#cb1-30" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC T cells&quot;</span>,</span>
<span id="cb1-31"><a href="#cb1-31" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,</span>
<span id="cb1-32"><a href="#cb1-32" tabindex="-1"></a>  <span class="at">api_key =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>),</span>
<span id="cb1-33"><a href="#cb1-33" tabindex="-1"></a>  <span class="at">top_gene_count =</span> <span class="dv">10</span></span>
<span id="cb1-34"><a href="#cb1-34" tabindex="-1"></a>)</span>
<span id="cb1-35"><a href="#cb1-35" tabindex="-1"></a></span>
<span id="cb1-36"><a href="#cb1-36" tabindex="-1"></a><span class="co"># Step 6: Add T cell subtypes back to original object</span></span>
<span id="cb1-37"><a href="#cb1-37" tabindex="-1"></a>t_cell_barcodes <span class="ot">&lt;-</span> <span class="fu">WhichCells</span>(t_cells)</span>
<span id="cb1-38"><a href="#cb1-38" tabindex="-1"></a>seurat_obj<span class="sc">$</span>detailed_celltype <span class="ot">&lt;-</span> seurat_obj<span class="sc">$</span>high_level_celltype</span>
<span id="cb1-39"><a href="#cb1-39" tabindex="-1"></a>seurat_obj<span class="sc">$</span>detailed_celltype[t_cell_barcodes] <span class="ot">&lt;-</span> plyr<span class="sc">::</span><span class="fu">mapvalues</span>(</span>
<span id="cb1-40"><a href="#cb1-40" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">as.character</span>(<span class="fu">Idents</span>(t_cells)),</span>
<span id="cb1-41"><a href="#cb1-41" tabindex="-1"></a>  <span class="at">from =</span> <span class="fu">names</span>(t_cell_subtypes),</span>
<span id="cb1-42"><a href="#cb1-42" tabindex="-1"></a>  <span class="at">to =</span> <span class="fu">paste0</span>(<span class="st">&quot;T cells: &quot;</span>, t_cell_subtypes)</span>
<span id="cb1-43"><a href="#cb1-43" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="validating-hierarchical-annotations" class="section level3">
<h3>Validating Hierarchical Annotations</h3>
<p>After creating hierarchical annotations, it’s important to validate
the consistency between levels:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># Create a simple function to check parent-child consistency</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>validate_hierarchy <span class="ot">&lt;-</span> <span class="cf">function</span>(high_level, detailed_level) {</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>  <span class="co"># Extract parent type from detailed annotation (before the colon)</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>  parent_from_detailed <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="fu">strsplit</span>(detailed_level, <span class="st">&quot;: &quot;</span>), <span class="cf">function</span>(x) x[<span class="dv">1</span>])</span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>  <span class="co"># Check if parent matches high-level annotation</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a>  consistent <span class="ot">&lt;-</span> parent_from_detailed <span class="sc">==</span> high_level</span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a>  <span class="co"># Return consistency check results</span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a>  <span class="fu">data.frame</span>(</span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a>    <span class="at">high_level =</span> high_level,</span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a>    <span class="at">detailed_level =</span> detailed_level,</span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a>    <span class="at">consistent =</span> consistent</span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a>  )</span>
<span id="cb2-15"><a href="#cb2-15" tabindex="-1"></a>}</span>
<span id="cb2-16"><a href="#cb2-16" tabindex="-1"></a></span>
<span id="cb2-17"><a href="#cb2-17" tabindex="-1"></a><span class="co"># Apply validation</span></span>
<span id="cb2-18"><a href="#cb2-18" tabindex="-1"></a>hierarchy_validation <span class="ot">&lt;-</span> <span class="fu">validate_hierarchy</span>(</span>
<span id="cb2-19"><a href="#cb2-19" tabindex="-1"></a>  seurat_obj<span class="sc">$</span>high_level_celltype,</span>
<span id="cb2-20"><a href="#cb2-20" tabindex="-1"></a>  seurat_obj<span class="sc">$</span>detailed_celltype</span>
<span id="cb2-21"><a href="#cb2-21" tabindex="-1"></a>)</span>
<span id="cb2-22"><a href="#cb2-22" tabindex="-1"></a></span>
<span id="cb2-23"><a href="#cb2-23" tabindex="-1"></a><span class="co"># Identify inconsistencies</span></span>
<span id="cb2-24"><a href="#cb2-24" tabindex="-1"></a>inconsistencies <span class="ot">&lt;-</span> hierarchy_validation[<span class="sc">!</span>hierarchy_validation<span class="sc">$</span>consistent, ]</span>
<span id="cb2-25"><a href="#cb2-25" tabindex="-1"></a><span class="fu">print</span>(inconsistencies)</span></code></pre></div>
</div>
</div>
<div id="handling-noisy-input-data" class="section level2">
<h2>Handling Noisy Input Data</h2>
<div id="strategies-for-noisy-marker-genes" class="section level3">
<h3>Strategies for Noisy Marker Genes</h3>
<p>Real-world scRNA-seq data often contains noise. Here are practical
strategies for handling noisy input:</p>
<div id="adjust-the-top_gene_count-parameter" class="section level4">
<h4>1. Adjust the top_gene_count parameter</h4>
<p>For noisy datasets, using fewer top genes can help focus on the
strongest signals:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># For noisy data, use fewer top genes</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>results_fewer_genes <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>  <span class="at">input =</span> marker_data,  <span class="co"># Your marker gene data</span></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>  <span class="at">api_key =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>),</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>  <span class="at">top_gene_count =</span> <span class="dv">5</span>  <span class="co"># Use fewer genes to focus on strongest signals</span></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="apply-stricter-filtering-for-marker-genes" class="section level4">
<h4>2. Apply stricter filtering for marker genes</h4>
<p>Pre-filtering marker genes with stricter thresholds can improve
annotation quality:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># Apply stricter filtering to marker genes</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>filtered_markers <span class="ot">&lt;-</span> marker_data <span class="sc">%&gt;%</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>  <span class="fu">filter</span>(p_val_adj <span class="sc">&lt;</span> <span class="fl">0.01</span>, avg_log2FC <span class="sc">&gt;</span> <span class="fl">1.0</span>)  <span class="co"># Stricter thresholds</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="co"># Annotate with filtered markers</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>results_filtered <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>  <span class="at">input =</span> filtered_markers,</span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,</span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a>  <span class="at">api_key =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>)</span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="use-multi-model-consensus" class="section level4">
<h4>3. Use multi-model consensus</h4>
<p>The consensus approach can help overcome noise by combining
predictions from multiple models:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># Set up API keys</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>api_keys <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>  <span class="at">anthropic =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>),</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a>  <span class="at">openai =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;OPENAI_API_KEY&quot;</span>),</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>  <span class="at">gemini =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;GEMINI_API_KEY&quot;</span>)</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>)</span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a><span class="co"># Define multiple models to use</span></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a>models <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a>  <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,</span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a>  <span class="st">&quot;gpt-5&quot;</span>,</span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a>  <span class="st">&quot;gemini-1.5-pro&quot;</span></span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a>)</span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a></span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a><span class="co"># Create consensus using interactive_consensus_annotation</span></span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a>consensus_results <span class="ot">&lt;-</span> <span class="fu">interactive_consensus_annotation</span>(</span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a>  <span class="at">input =</span> marker_data,  <span class="co"># Your marker gene data</span></span>
<span id="cb5-18"><a href="#cb5-18" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb5-19"><a href="#cb5-19" tabindex="-1"></a>  <span class="at">models =</span> models,</span>
<span id="cb5-20"><a href="#cb5-20" tabindex="-1"></a>  <span class="at">api_keys =</span> api_keys,</span>
<span id="cb5-21"><a href="#cb5-21" tabindex="-1"></a>  <span class="at">controversy_threshold =</span> <span class="fl">0.7</span>,</span>
<span id="cb5-22"><a href="#cb5-22" tabindex="-1"></a>  <span class="at">entropy_threshold =</span> <span class="fl">1.0</span>,</span>
<span id="cb5-23"><a href="#cb5-23" tabindex="-1"></a>  <span class="at">consensus_check_model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span></span>
<span id="cb5-24"><a href="#cb5-24" tabindex="-1"></a>)</span></code></pre></div>
</div>
</div>
<div id="handling-data-with-batch-effects" class="section level3">
<h3>Handling Data with Batch Effects</h3>
<p>When working with data affected by batch effects, you can:</p>
<div id="use-the-consensus-approach-with-a-lower-controversy-threshold" class="section level4">
<h4>1. Use the consensus approach with a lower controversy
threshold</h4>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># For data with batch effects, use consensus with lower threshold</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>batch_consensus <span class="ot">&lt;-</span> <span class="fu">interactive_consensus_annotation</span>(</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>  <span class="at">input =</span> marker_data,  <span class="co"># Your marker gene data with batch effects</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;mouse brain&quot;</span>,</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>  <span class="at">models =</span> <span class="fu">c</span>(<span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>, <span class="st">&quot;gpt-5&quot;</span>, <span class="st">&quot;gemini-1.5-pro&quot;</span>),</span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>  <span class="at">api_keys =</span> api_keys,</span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>  <span class="at">controversy_threshold =</span> <span class="fl">0.4</span>,  <span class="co"># Lower threshold to discuss more clusters</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a>  <span class="at">entropy_threshold =</span> <span class="fl">0.8</span>  <span class="co"># Lower entropy threshold</span></span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="include-batch-information-in-the-tissue-context" class="section level4">
<h4>2. Include batch information in the tissue context</h4>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="co"># Include batch information in the tissue context</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>batch_aware_results <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>  <span class="at">input =</span> marker_data,  <span class="co"># Your marker gene data with batch effects</span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;mouse brain with technical batch effects&quot;</span>,  <span class="co"># Include batch context</span></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,</span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a>  <span class="at">api_key =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>)</span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a>)</span></code></pre></div>
</div>
</div>
</div>
<div id="incorporating-domain-knowledge" class="section level2">
<h2>Incorporating Domain Knowledge</h2>
<div id="using-tissue-context" class="section level3">
<h3>Using Tissue Context</h3>
<p>One of the key features of mLLMCelltype is the ability to incorporate
domain knowledge through the <code>tissue_name</code> parameter. This
provides important context to the LLM:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co"># Basic annotation without specific tissue context</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>basic_results <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>  <span class="at">input =</span> marker_data,</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human sample&quot;</span>,  <span class="co"># Generic context</span></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,</span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a>  <span class="at">api_key =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>)</span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a>)</span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a><span class="co"># Annotation with specific tissue context</span></span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a>specific_results <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a>  <span class="at">input =</span> marker_data,</span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human fetal liver at 20 weeks gestation&quot;</span>,  <span class="co"># Detailed context</span></span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,</span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a>  <span class="at">api_key =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>)</span>
<span id="cb8-15"><a href="#cb8-15" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="creating-custom-prompts" class="section level3">
<h3>Creating Custom Prompts</h3>
<p>For advanced use cases, you can create and modify the annotation
prompt directly:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="co"># Create a custom annotation prompt</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>custom_prompt <span class="ot">&lt;-</span> <span class="fu">create_annotation_prompt</span>(</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>  <span class="at">input =</span> marker_data,</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>  <span class="at">top_gene_count =</span> <span class="dv">10</span></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a>)</span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a><span class="co"># Modify the prompt to include additional context</span></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a>modified_prompt <span class="ot">&lt;-</span> <span class="fu">paste0</span>(</span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a>  custom_prompt<span class="sc">$</span>prompt,</span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a>  <span class="st">&quot;</span><span class="sc">\n\n</span><span class="st">Additional context: This sample is from a patient with rheumatoid arthritis. &quot;</span>,</span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a>  <span class="st">&quot;Previous studies have identified activated T cells, B cells, and CXCR4-high monocytes in this condition.&quot;</span></span>
<span id="cb9-13"><a href="#cb9-13" tabindex="-1"></a>)</span>
<span id="cb9-14"><a href="#cb9-14" tabindex="-1"></a></span>
<span id="cb9-15"><a href="#cb9-15" tabindex="-1"></a><span class="co"># Use the modified prompt directly</span></span>
<span id="cb9-16"><a href="#cb9-16" tabindex="-1"></a>custom_results <span class="ot">&lt;-</span> <span class="fu">get_model_response</span>(</span>
<span id="cb9-17"><a href="#cb9-17" tabindex="-1"></a>  <span class="at">prompt =</span> modified_prompt,</span>
<span id="cb9-18"><a href="#cb9-18" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,</span>
<span id="cb9-19"><a href="#cb9-19" tabindex="-1"></a>  <span class="at">api_key =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>)</span>
<span id="cb9-20"><a href="#cb9-20" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="combining-with-external-resources" class="section level3">
<h3>Combining with External Resources</h3>
<p>You can enhance your annotation workflow by combining mLLMCelltype
with other R packages and resources:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a><span class="co"># Example: Using CellMarker database information to validate annotations</span></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a><span class="co"># This is a conceptual example - implementation would depend on your specific needs</span></span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a></span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a><span class="co"># 1. Get annotations with mLLMCelltype</span></span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a>annotations <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a>  <span class="at">input =</span> marker_data,</span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb10-11"><a href="#cb10-11" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,</span>
<span id="cb10-12"><a href="#cb10-12" tabindex="-1"></a>  <span class="at">api_key =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>)</span>
<span id="cb10-13"><a href="#cb10-13" tabindex="-1"></a>)</span>
<span id="cb10-14"><a href="#cb10-14" tabindex="-1"></a></span>
<span id="cb10-15"><a href="#cb10-15" tabindex="-1"></a><span class="co"># 2. Compare with known marker genes (conceptual)</span></span>
<span id="cb10-16"><a href="#cb10-16" tabindex="-1"></a><span class="co"># In a real workflow, you would query a database or use a reference dataset</span></span>
<span id="cb10-17"><a href="#cb10-17" tabindex="-1"></a>known_markers <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb10-18"><a href="#cb10-18" tabindex="-1"></a>  <span class="st">&quot;T cells&quot;</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;CD3D&quot;</span>, <span class="st">&quot;CD3E&quot;</span>, <span class="st">&quot;CD3G&quot;</span>),</span>
<span id="cb10-19"><a href="#cb10-19" tabindex="-1"></a>  <span class="st">&quot;B cells&quot;</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;CD19&quot;</span>, <span class="st">&quot;MS4A1&quot;</span>, <span class="st">&quot;CD79A&quot;</span>),</span>
<span id="cb10-20"><a href="#cb10-20" tabindex="-1"></a>  <span class="st">&quot;Monocytes&quot;</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;CD14&quot;</span>, <span class="st">&quot;LYZ&quot;</span>, <span class="st">&quot;CSF1R&quot;</span>)</span>
<span id="cb10-21"><a href="#cb10-21" tabindex="-1"></a>)</span>
<span id="cb10-22"><a href="#cb10-22" tabindex="-1"></a></span>
<span id="cb10-23"><a href="#cb10-23" tabindex="-1"></a><span class="co"># 3. Validate annotations against known markers</span></span>
<span id="cb10-24"><a href="#cb10-24" tabindex="-1"></a><span class="co"># This is a simplified example of how you might validate annotations</span></span>
<span id="cb10-25"><a href="#cb10-25" tabindex="-1"></a>validate_annotations <span class="ot">&lt;-</span> <span class="cf">function</span>(annotations, marker_data, known_markers) {</span>
<span id="cb10-26"><a href="#cb10-26" tabindex="-1"></a>  validation_results <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb10-27"><a href="#cb10-27" tabindex="-1"></a></span>
<span id="cb10-28"><a href="#cb10-28" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(annotations)) {</span>
<span id="cb10-29"><a href="#cb10-29" tabindex="-1"></a>    cluster_id <span class="ot">&lt;-</span> i</span>
<span id="cb10-30"><a href="#cb10-30" tabindex="-1"></a>    predicted_type <span class="ot">&lt;-</span> annotations[i]</span>
<span id="cb10-31"><a href="#cb10-31" tabindex="-1"></a></span>
<span id="cb10-32"><a href="#cb10-32" tabindex="-1"></a>    <span class="co"># Get markers for this cluster</span></span>
<span id="cb10-33"><a href="#cb10-33" tabindex="-1"></a>    cluster_markers <span class="ot">&lt;-</span> marker_data <span class="sc">%&gt;%</span></span>
<span id="cb10-34"><a href="#cb10-34" tabindex="-1"></a>      <span class="fu">filter</span>(cluster <span class="sc">==</span> cluster_id) <span class="sc">%&gt;%</span></span>
<span id="cb10-35"><a href="#cb10-35" tabindex="-1"></a>      <span class="fu">arrange</span>(<span class="fu">desc</span>(avg_log2FC)) <span class="sc">%&gt;%</span></span>
<span id="cb10-36"><a href="#cb10-36" tabindex="-1"></a>      <span class="fu">pull</span>(gene) <span class="sc">%&gt;%</span></span>
<span id="cb10-37"><a href="#cb10-37" tabindex="-1"></a>      <span class="fu">head</span>(<span class="dv">20</span>)</span>
<span id="cb10-38"><a href="#cb10-38" tabindex="-1"></a></span>
<span id="cb10-39"><a href="#cb10-39" tabindex="-1"></a>    <span class="co"># Check overlap with known markers for this cell type</span></span>
<span id="cb10-40"><a href="#cb10-40" tabindex="-1"></a>    <span class="cf">if</span> (predicted_type <span class="sc">%in%</span> <span class="fu">names</span>(known_markers)) {</span>
<span id="cb10-41"><a href="#cb10-41" tabindex="-1"></a>      expected_markers <span class="ot">&lt;-</span> known_markers[[predicted_type]]</span>
<span id="cb10-42"><a href="#cb10-42" tabindex="-1"></a>      overlap <span class="ot">&lt;-</span> <span class="fu">intersect</span>(cluster_markers, expected_markers)</span>
<span id="cb10-43"><a href="#cb10-43" tabindex="-1"></a></span>
<span id="cb10-44"><a href="#cb10-44" tabindex="-1"></a>      validation_results[[i]] <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb10-45"><a href="#cb10-45" tabindex="-1"></a>        <span class="at">cluster =</span> cluster_id,</span>
<span id="cb10-46"><a href="#cb10-46" tabindex="-1"></a>        <span class="at">predicted_type =</span> predicted_type,</span>
<span id="cb10-47"><a href="#cb10-47" tabindex="-1"></a>        <span class="at">overlap_count =</span> <span class="fu">length</span>(overlap),</span>
<span id="cb10-48"><a href="#cb10-48" tabindex="-1"></a>        <span class="at">overlap_genes =</span> <span class="fu">paste</span>(overlap, <span class="at">collapse =</span> <span class="st">&quot;, &quot;</span>),</span>
<span id="cb10-49"><a href="#cb10-49" tabindex="-1"></a>        <span class="at">confidence =</span> <span class="fu">length</span>(overlap) <span class="sc">/</span> <span class="fu">length</span>(expected_markers)</span>
<span id="cb10-50"><a href="#cb10-50" tabindex="-1"></a>      )</span>
<span id="cb10-51"><a href="#cb10-51" tabindex="-1"></a>    } <span class="cf">else</span> {</span>
<span id="cb10-52"><a href="#cb10-52" tabindex="-1"></a>      validation_results[[i]] <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb10-53"><a href="#cb10-53" tabindex="-1"></a>        <span class="at">cluster =</span> cluster_id,</span>
<span id="cb10-54"><a href="#cb10-54" tabindex="-1"></a>        <span class="at">predicted_type =</span> predicted_type,</span>
<span id="cb10-55"><a href="#cb10-55" tabindex="-1"></a>        <span class="at">overlap_count =</span> <span class="dv">0</span>,</span>
<span id="cb10-56"><a href="#cb10-56" tabindex="-1"></a>        <span class="at">overlap_genes =</span> <span class="st">&quot;&quot;</span>,</span>
<span id="cb10-57"><a href="#cb10-57" tabindex="-1"></a>        <span class="at">confidence =</span> <span class="dv">0</span></span>
<span id="cb10-58"><a href="#cb10-58" tabindex="-1"></a>      )</span>
<span id="cb10-59"><a href="#cb10-59" tabindex="-1"></a>    }</span>
<span id="cb10-60"><a href="#cb10-60" tabindex="-1"></a>  }</span>
<span id="cb10-61"><a href="#cb10-61" tabindex="-1"></a></span>
<span id="cb10-62"><a href="#cb10-62" tabindex="-1"></a>  <span class="fu">return</span>(validation_results)</span>
<span id="cb10-63"><a href="#cb10-63" tabindex="-1"></a>}</span>
<span id="cb10-64"><a href="#cb10-64" tabindex="-1"></a></span>
<span id="cb10-65"><a href="#cb10-65" tabindex="-1"></a><span class="co"># This is a conceptual example of how you might validate annotations</span></span>
<span id="cb10-66"><a href="#cb10-66" tabindex="-1"></a><span class="co"># validation_results &lt;- validate_annotations(annotations, marker_data, known_markers)</span></span></code></pre></div>
</div>
</div>
<div id="practical-case-studies" class="section level2">
<h2>Practical Case Studies</h2>
<div id="case-study-1-pbmc-dataset-analysis" class="section level3">
<h3>Case Study 1: PBMC Dataset Analysis</h3>
<p>This example demonstrates a complete workflow for analyzing a PBMC
dataset:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="fu">library</span>(mLLMCelltype)</span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a></span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a><span class="co"># Load example PBMC data</span></span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a><span class="co"># In a real workflow, you would use your own data</span></span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;pbmc_small&quot;</span>)  <span class="co"># Example dataset from Seurat</span></span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a></span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a><span class="co"># Find marker genes</span></span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a>pbmc_markers <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(pbmc_small,</span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a>                              <span class="at">only.pos =</span> <span class="cn">TRUE</span>,</span>
<span id="cb11-13"><a href="#cb11-13" tabindex="-1"></a>                              <span class="at">min.pct =</span> <span class="fl">0.25</span>,</span>
<span id="cb11-14"><a href="#cb11-14" tabindex="-1"></a>                              <span class="at">logfc.threshold =</span> <span class="fl">0.25</span>)</span>
<span id="cb11-15"><a href="#cb11-15" tabindex="-1"></a></span>
<span id="cb11-16"><a href="#cb11-16" tabindex="-1"></a><span class="co"># Set up API keys</span></span>
<span id="cb11-17"><a href="#cb11-17" tabindex="-1"></a>api_keys <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb11-18"><a href="#cb11-18" tabindex="-1"></a>  <span class="at">anthropic =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>),</span>
<span id="cb11-19"><a href="#cb11-19" tabindex="-1"></a>  <span class="at">openai =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;OPENAI_API_KEY&quot;</span>),</span>
<span id="cb11-20"><a href="#cb11-20" tabindex="-1"></a>  <span class="at">gemini =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;GEMINI_API_KEY&quot;</span>)</span>
<span id="cb11-21"><a href="#cb11-21" tabindex="-1"></a>)</span>
<span id="cb11-22"><a href="#cb11-22" tabindex="-1"></a></span>
<span id="cb11-23"><a href="#cb11-23" tabindex="-1"></a><span class="co"># Use consensus annotation</span></span>
<span id="cb11-24"><a href="#cb11-24" tabindex="-1"></a>consensus_results <span class="ot">&lt;-</span> <span class="fu">interactive_consensus_annotation</span>(</span>
<span id="cb11-25"><a href="#cb11-25" tabindex="-1"></a>  <span class="at">input =</span> pbmc_markers,</span>
<span id="cb11-26"><a href="#cb11-26" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb11-27"><a href="#cb11-27" tabindex="-1"></a>  <span class="at">models =</span> <span class="fu">c</span>(<span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>, <span class="st">&quot;gpt-5&quot;</span>, <span class="st">&quot;gemini-1.5-pro&quot;</span>),</span>
<span id="cb11-28"><a href="#cb11-28" tabindex="-1"></a>  <span class="at">api_keys =</span> api_keys,</span>
<span id="cb11-29"><a href="#cb11-29" tabindex="-1"></a>  <span class="at">controversy_threshold =</span> <span class="fl">0.7</span>,</span>
<span id="cb11-30"><a href="#cb11-30" tabindex="-1"></a>  <span class="at">entropy_threshold =</span> <span class="fl">1.0</span>,</span>
<span id="cb11-31"><a href="#cb11-31" tabindex="-1"></a>  <span class="at">consensus_check_model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span></span>
<span id="cb11-32"><a href="#cb11-32" tabindex="-1"></a>)</span>
<span id="cb11-33"><a href="#cb11-33" tabindex="-1"></a></span>
<span id="cb11-34"><a href="#cb11-34" tabindex="-1"></a><span class="co"># Add results to Seurat object</span></span>
<span id="cb11-35"><a href="#cb11-35" tabindex="-1"></a>pbmc_small<span class="sc">$</span>cell_type <span class="ot">&lt;-</span> plyr<span class="sc">::</span><span class="fu">mapvalues</span>(</span>
<span id="cb11-36"><a href="#cb11-36" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">as.character</span>(<span class="fu">Idents</span>(pbmc_small)),</span>
<span id="cb11-37"><a href="#cb11-37" tabindex="-1"></a>  <span class="at">from =</span> <span class="fu">names</span>(consensus_results<span class="sc">$</span>final_annotations),</span>
<span id="cb11-38"><a href="#cb11-38" tabindex="-1"></a>  <span class="at">to =</span> consensus_results<span class="sc">$</span>final_annotations</span>
<span id="cb11-39"><a href="#cb11-39" tabindex="-1"></a>)</span>
<span id="cb11-40"><a href="#cb11-40" tabindex="-1"></a></span>
<span id="cb11-41"><a href="#cb11-41" tabindex="-1"></a><span class="co"># Visualize results</span></span>
<span id="cb11-42"><a href="#cb11-42" tabindex="-1"></a><span class="co"># In a real workflow, you would create a UMAP or t-SNE plot</span></span>
<span id="cb11-43"><a href="#cb11-43" tabindex="-1"></a><span class="co"># DimPlot(pbmc_small, group.by = &quot;cell_type&quot;, label = TRUE) +</span></span>
<span id="cb11-44"><a href="#cb11-44" tabindex="-1"></a><span class="co">#   ggtitle(&quot;PBMC Cell Types&quot;)</span></span></code></pre></div>
</div>
<div id="case-study-2-identifying-rare-cell-types" class="section level3">
<h3>Case Study 2: Identifying Rare Cell Types</h3>
<p>When working with datasets containing rare cell populations, you can
adjust parameters to improve detection:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="co"># For rare cell types, use these strategies:</span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a></span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a><span class="co"># 1. Increase the number of marker genes considered</span></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a>rare_cell_annotation <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a>  <span class="at">input =</span> marker_data,  <span class="co"># Your marker gene data</span></span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human bone marrow&quot;</span>,</span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,</span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a>  <span class="at">api_key =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>),</span>
<span id="cb12-9"><a href="#cb12-9" tabindex="-1"></a>  <span class="at">top_gene_count =</span> <span class="dv">20</span>  <span class="co"># Use more genes for rare cell types</span></span>
<span id="cb12-10"><a href="#cb12-10" tabindex="-1"></a>)</span>
<span id="cb12-11"><a href="#cb12-11" tabindex="-1"></a></span>
<span id="cb12-12"><a href="#cb12-12" tabindex="-1"></a><span class="co"># 2. Use consensus with lower thresholds to discuss more clusters</span></span>
<span id="cb12-13"><a href="#cb12-13" tabindex="-1"></a>rare_cell_consensus <span class="ot">&lt;-</span> <span class="fu">interactive_consensus_annotation</span>(</span>
<span id="cb12-14"><a href="#cb12-14" tabindex="-1"></a>  <span class="at">input =</span> marker_data,  <span class="co"># Your marker gene data</span></span>
<span id="cb12-15"><a href="#cb12-15" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human bone marrow&quot;</span>,</span>
<span id="cb12-16"><a href="#cb12-16" tabindex="-1"></a>  <span class="at">models =</span> <span class="fu">c</span>(<span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>, <span class="st">&quot;gpt-5&quot;</span>, <span class="st">&quot;gemini-1.5-pro&quot;</span>),</span>
<span id="cb12-17"><a href="#cb12-17" tabindex="-1"></a>  <span class="at">api_keys =</span> api_keys,</span>
<span id="cb12-18"><a href="#cb12-18" tabindex="-1"></a>  <span class="at">controversy_threshold =</span> <span class="fl">0.4</span>,  <span class="co"># Lower threshold to discuss more clusters</span></span>
<span id="cb12-19"><a href="#cb12-19" tabindex="-1"></a>  <span class="at">entropy_threshold =</span> <span class="fl">0.8</span>,  <span class="co"># Lower entropy threshold</span></span>
<span id="cb12-20"><a href="#cb12-20" tabindex="-1"></a>  <span class="at">consensus_check_model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span></span>
<span id="cb12-21"><a href="#cb12-21" tabindex="-1"></a>)</span>
<span id="cb12-22"><a href="#cb12-22" tabindex="-1"></a></span>
<span id="cb12-23"><a href="#cb12-23" tabindex="-1"></a><span class="co"># 3. Provide more specific tissue context</span></span>
<span id="cb12-24"><a href="#cb12-24" tabindex="-1"></a>specific_annotation <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb12-25"><a href="#cb12-25" tabindex="-1"></a>  <span class="at">input =</span> marker_data,  <span class="co"># Your marker gene data</span></span>
<span id="cb12-26"><a href="#cb12-26" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human bone marrow with expected rare plasma cells and basophils&quot;</span>,</span>
<span id="cb12-27"><a href="#cb12-27" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,</span>
<span id="cb12-28"><a href="#cb12-28" tabindex="-1"></a>  <span class="at">api_key =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>)</span>
<span id="cb12-29"><a href="#cb12-29" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="case-study-3-cross-species-comparison" class="section level3">
<h3>Case Study 3: Cross-Species Comparison</h3>
<p>mLLMCelltype can be used to compare cell types across different
species:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="co"># Example workflow for cross-species comparison</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a><span class="co"># 1. Annotate human and mouse datasets separately</span></span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a><span class="co"># (Assuming you have marker data for both species)</span></span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a>human_annotations <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a>  <span class="at">input =</span> human_marker_data,  <span class="co"># Your human marker data</span></span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human brain cortex&quot;</span>,</span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,</span>
<span id="cb13-9"><a href="#cb13-9" tabindex="-1"></a>  <span class="at">api_key =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>)</span>
<span id="cb13-10"><a href="#cb13-10" tabindex="-1"></a>)</span>
<span id="cb13-11"><a href="#cb13-11" tabindex="-1"></a></span>
<span id="cb13-12"><a href="#cb13-12" tabindex="-1"></a>mouse_annotations <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb13-13"><a href="#cb13-13" tabindex="-1"></a>  <span class="at">input =</span> mouse_marker_data,  <span class="co"># Your mouse marker data</span></span>
<span id="cb13-14"><a href="#cb13-14" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;mouse brain cortex&quot;</span>,</span>
<span id="cb13-15"><a href="#cb13-15" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,</span>
<span id="cb13-16"><a href="#cb13-16" tabindex="-1"></a>  <span class="at">api_key =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>)</span>
<span id="cb13-17"><a href="#cb13-17" tabindex="-1"></a>)</span>
<span id="cb13-18"><a href="#cb13-18" tabindex="-1"></a></span>
<span id="cb13-19"><a href="#cb13-19" tabindex="-1"></a><span class="co"># 2. Compare annotations</span></span>
<span id="cb13-20"><a href="#cb13-20" tabindex="-1"></a><span class="co"># This is a conceptual example - in a real workflow, you would:</span></span>
<span id="cb13-21"><a href="#cb13-21" tabindex="-1"></a><span class="co"># - Map annotations to Seurat objects</span></span>
<span id="cb13-22"><a href="#cb13-22" tabindex="-1"></a><span class="co"># - Calculate proportions</span></span>
<span id="cb13-23"><a href="#cb13-23" tabindex="-1"></a><span class="co"># - Create comparison visualizations</span></span>
<span id="cb13-24"><a href="#cb13-24" tabindex="-1"></a><span class="co"># - Identify conserved and species-specific cell types</span></span>
<span id="cb13-25"><a href="#cb13-25" tabindex="-1"></a></span>
<span id="cb13-26"><a href="#cb13-26" tabindex="-1"></a><span class="co"># Example comparison function (conceptual)</span></span>
<span id="cb13-27"><a href="#cb13-27" tabindex="-1"></a>compare_species_annotations <span class="ot">&lt;-</span> <span class="cf">function</span>(human_annotations, mouse_annotations) {</span>
<span id="cb13-28"><a href="#cb13-28" tabindex="-1"></a>  <span class="co"># Get unique cell types from both species</span></span>
<span id="cb13-29"><a href="#cb13-29" tabindex="-1"></a>  human_types <span class="ot">&lt;-</span> <span class="fu">unique</span>(human_annotations)</span>
<span id="cb13-30"><a href="#cb13-30" tabindex="-1"></a>  mouse_types <span class="ot">&lt;-</span> <span class="fu">unique</span>(mouse_annotations)</span>
<span id="cb13-31"><a href="#cb13-31" tabindex="-1"></a></span>
<span id="cb13-32"><a href="#cb13-32" tabindex="-1"></a>  <span class="co"># Find common cell types</span></span>
<span id="cb13-33"><a href="#cb13-33" tabindex="-1"></a>  common_types <span class="ot">&lt;-</span> <span class="fu">intersect</span>(human_types, mouse_types)</span>
<span id="cb13-34"><a href="#cb13-34" tabindex="-1"></a></span>
<span id="cb13-35"><a href="#cb13-35" tabindex="-1"></a>  <span class="co"># Find species-specific cell types</span></span>
<span id="cb13-36"><a href="#cb13-36" tabindex="-1"></a>  human_specific <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(human_types, mouse_types)</span>
<span id="cb13-37"><a href="#cb13-37" tabindex="-1"></a>  mouse_specific <span class="ot">&lt;-</span> <span class="fu">setdiff</span>(mouse_types, human_types)</span>
<span id="cb13-38"><a href="#cb13-38" tabindex="-1"></a></span>
<span id="cb13-39"><a href="#cb13-39" tabindex="-1"></a>  <span class="co"># Return comparison results</span></span>
<span id="cb13-40"><a href="#cb13-40" tabindex="-1"></a>  <span class="fu">list</span>(</span>
<span id="cb13-41"><a href="#cb13-41" tabindex="-1"></a>    <span class="at">common_types =</span> common_types,</span>
<span id="cb13-42"><a href="#cb13-42" tabindex="-1"></a>    <span class="at">human_specific =</span> human_specific,</span>
<span id="cb13-43"><a href="#cb13-43" tabindex="-1"></a>    <span class="at">mouse_specific =</span> mouse_specific</span>
<span id="cb13-44"><a href="#cb13-44" tabindex="-1"></a>  )</span>
<span id="cb13-45"><a href="#cb13-45" tabindex="-1"></a>}</span>
<span id="cb13-46"><a href="#cb13-46" tabindex="-1"></a></span>
<span id="cb13-47"><a href="#cb13-47" tabindex="-1"></a><span class="co"># This is a conceptual example</span></span>
<span id="cb13-48"><a href="#cb13-48" tabindex="-1"></a><span class="co"># comparison &lt;- compare_species_annotations(human_annotations, mouse_annotations)</span></span></code></pre></div>
</div>
</div>
<div id="performance-considerations" class="section level2">
<h2>Performance Considerations</h2>
<div id="api-cost-management" class="section level3">
<h3>API Cost Management</h3>
<p>When using mLLMCelltype, it’s important to consider the costs
associated with API calls to different LLM providers:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="co"># Example of cost-efficient model selection</span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="co"># Choose models based on your specific needs and budget</span></span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a></span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a><span class="co"># For initial exploration or smaller datasets</span></span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a><span class="co"># Use more affordable models</span></span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a>affordable_annotation <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb14-7"><a href="#cb14-7" tabindex="-1"></a>  <span class="at">input =</span> marker_data,  <span class="co"># Your marker gene data</span></span>
<span id="cb14-8"><a href="#cb14-8" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb14-9"><a href="#cb14-9" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;claude-haiku-4-20250514&quot;</span>,  <span class="co"># More affordable model</span></span>
<span id="cb14-10"><a href="#cb14-10" tabindex="-1"></a>  <span class="at">api_key =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>)</span>
<span id="cb14-11"><a href="#cb14-11" tabindex="-1"></a>)</span>
<span id="cb14-12"><a href="#cb14-12" tabindex="-1"></a></span>
<span id="cb14-13"><a href="#cb14-13" tabindex="-1"></a><span class="co"># For final analysis or challenging datasets</span></span>
<span id="cb14-14"><a href="#cb14-14" tabindex="-1"></a><span class="co"># Use larger models</span></span>
<span id="cb14-15"><a href="#cb14-15" tabindex="-1"></a>premium_annotation <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb14-16"><a href="#cb14-16" tabindex="-1"></a>  <span class="at">input =</span> marker_data,  <span class="co"># Your marker gene data</span></span>
<span id="cb14-17"><a href="#cb14-17" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb14-18"><a href="#cb14-18" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,  <span class="co"># Larger model</span></span>
<span id="cb14-19"><a href="#cb14-19" tabindex="-1"></a>  <span class="at">api_key =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>)</span>
<span id="cb14-20"><a href="#cb14-20" tabindex="-1"></a>)</span>
<span id="cb14-21"><a href="#cb14-21" tabindex="-1"></a></span>
<span id="cb14-22"><a href="#cb14-22" tabindex="-1"></a><span class="co"># Use OpenRouter for access to free models</span></span>
<span id="cb14-23"><a href="#cb14-23" tabindex="-1"></a>openrouter_annotation <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb14-24"><a href="#cb14-24" tabindex="-1"></a>  <span class="at">input =</span> marker_data,  <span class="co"># Your marker gene data</span></span>
<span id="cb14-25"><a href="#cb14-25" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb14-26"><a href="#cb14-26" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;meta-llama/llama-3.3-70b-instruct:free&quot;</span>,  <span class="co"># Free model via OpenRouter</span></span>
<span id="cb14-27"><a href="#cb14-27" tabindex="-1"></a>  <span class="at">api_key =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;OPENROUTER_API_KEY&quot;</span>)</span>
<span id="cb14-28"><a href="#cb14-28" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="optimizing-runtime" class="section level3">
<h3>Optimizing Runtime</h3>
<p>To optimize runtime when working with large datasets:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="co"># 1. Use caching with interactive_consensus_annotation</span></span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a>consensus_with_cache <span class="ot">&lt;-</span> <span class="fu">interactive_consensus_annotation</span>(</span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a>  <span class="at">input =</span> marker_data,  <span class="co"># Your marker gene data</span></span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a>  <span class="at">models =</span> <span class="fu">c</span>(<span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>, <span class="st">&quot;gpt-5&quot;</span>),</span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a>  <span class="at">api_keys =</span> api_keys,</span>
<span id="cb15-7"><a href="#cb15-7" tabindex="-1"></a>  <span class="at">use_cache =</span> <span class="cn">TRUE</span>,  <span class="co"># Enable caching</span></span>
<span id="cb15-8"><a href="#cb15-8" tabindex="-1"></a>  <span class="at">cache_dir =</span> <span class="cn">NULL</span>  <span class="co"># Uses default system cache directory</span></span>
<span id="cb15-9"><a href="#cb15-9" tabindex="-1"></a>)</span>
<span id="cb15-10"><a href="#cb15-10" tabindex="-1"></a></span>
<span id="cb15-11"><a href="#cb15-11" tabindex="-1"></a><span class="co"># 2. Process clusters in batches</span></span>
<span id="cb15-12"><a href="#cb15-12" tabindex="-1"></a><span class="co"># This is a conceptual example - implementation would depend on your workflow</span></span>
<span id="cb15-13"><a href="#cb15-13" tabindex="-1"></a>process_in_batches <span class="ot">&lt;-</span> <span class="cf">function</span>(marker_data, <span class="at">batch_size =</span> <span class="dv">5</span>) {</span>
<span id="cb15-14"><a href="#cb15-14" tabindex="-1"></a>  <span class="co"># Get unique clusters</span></span>
<span id="cb15-15"><a href="#cb15-15" tabindex="-1"></a>  clusters <span class="ot">&lt;-</span> <span class="fu">unique</span>(marker_data<span class="sc">$</span>cluster)</span>
<span id="cb15-16"><a href="#cb15-16" tabindex="-1"></a></span>
<span id="cb15-17"><a href="#cb15-17" tabindex="-1"></a>  <span class="co"># Process in batches</span></span>
<span id="cb15-18"><a href="#cb15-18" tabindex="-1"></a>  results <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb15-19"><a href="#cb15-19" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">seq</span>(<span class="dv">1</span>, <span class="fu">length</span>(clusters), <span class="at">by =</span> batch_size)) {</span>
<span id="cb15-20"><a href="#cb15-20" tabindex="-1"></a>    <span class="co"># Get current batch of clusters</span></span>
<span id="cb15-21"><a href="#cb15-21" tabindex="-1"></a>    batch_clusters <span class="ot">&lt;-</span> clusters[i<span class="sc">:</span><span class="fu">min</span>(i <span class="sc">+</span> batch_size <span class="sc">-</span> <span class="dv">1</span>, <span class="fu">length</span>(clusters))]</span>
<span id="cb15-22"><a href="#cb15-22" tabindex="-1"></a></span>
<span id="cb15-23"><a href="#cb15-23" tabindex="-1"></a>    <span class="co"># Filter marker data for current batch</span></span>
<span id="cb15-24"><a href="#cb15-24" tabindex="-1"></a>    batch_data <span class="ot">&lt;-</span> marker_data <span class="sc">%&gt;%</span> <span class="fu">filter</span>(cluster <span class="sc">%in%</span> batch_clusters)</span>
<span id="cb15-25"><a href="#cb15-25" tabindex="-1"></a></span>
<span id="cb15-26"><a href="#cb15-26" tabindex="-1"></a>    <span class="co"># Process batch</span></span>
<span id="cb15-27"><a href="#cb15-27" tabindex="-1"></a>    batch_results <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb15-28"><a href="#cb15-28" tabindex="-1"></a>      <span class="at">input =</span> batch_data,</span>
<span id="cb15-29"><a href="#cb15-29" tabindex="-1"></a>      <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb15-30"><a href="#cb15-30" tabindex="-1"></a>      <span class="at">model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,</span>
<span id="cb15-31"><a href="#cb15-31" tabindex="-1"></a>      <span class="at">api_key =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>)</span>
<span id="cb15-32"><a href="#cb15-32" tabindex="-1"></a>    )</span>
<span id="cb15-33"><a href="#cb15-33" tabindex="-1"></a></span>
<span id="cb15-34"><a href="#cb15-34" tabindex="-1"></a>    <span class="co"># Store results</span></span>
<span id="cb15-35"><a href="#cb15-35" tabindex="-1"></a>    results <span class="ot">&lt;-</span> <span class="fu">c</span>(results, batch_results)</span>
<span id="cb15-36"><a href="#cb15-36" tabindex="-1"></a>  }</span>
<span id="cb15-37"><a href="#cb15-37" tabindex="-1"></a></span>
<span id="cb15-38"><a href="#cb15-38" tabindex="-1"></a>  <span class="fu">return</span>(results)</span>
<span id="cb15-39"><a href="#cb15-39" tabindex="-1"></a>}</span>
<span id="cb15-40"><a href="#cb15-40" tabindex="-1"></a></span>
<span id="cb15-41"><a href="#cb15-41" tabindex="-1"></a><span class="co"># 3. Use faster models for initial exploration</span></span>
<span id="cb15-42"><a href="#cb15-42" tabindex="-1"></a>fast_annotation <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb15-43"><a href="#cb15-43" tabindex="-1"></a>  <span class="at">input =</span> marker_data,  <span class="co"># Your marker gene data</span></span>
<span id="cb15-44"><a href="#cb15-44" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb15-45"><a href="#cb15-45" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;claude-haiku-4-20250514&quot;</span>,  <span class="co"># Faster model</span></span>
<span id="cb15-46"><a href="#cb15-46" tabindex="-1"></a>  <span class="at">api_key =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>)</span>
<span id="cb15-47"><a href="#cb15-47" tabindex="-1"></a>)</span></code></pre></div>
</div>
</div>
<div id="advanced-customization" class="section level2">
<h2>Advanced Customization</h2>
<div id="custom-processing-functions" class="section level3">
<h3>Custom Processing Functions</h3>
<p>For advanced users, mLLMCelltype allows you to register custom
providers and models:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="co"># Define a custom processing function</span></span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a><span class="co"># This function must accept prompt, model, and api_key parameters</span></span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a>custom_process_fn <span class="ot">&lt;-</span> <span class="cf">function</span>(prompt, model, api_key) {</span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a>  <span class="co"># Custom implementation to process prompts and get responses</span></span>
<span id="cb16-5"><a href="#cb16-5" tabindex="-1"></a>  <span class="co"># This is a simplified example</span></span>
<span id="cb16-6"><a href="#cb16-6" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">&quot;Processing prompt with custom provider</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb16-7"><a href="#cb16-7" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">&quot;Model:&quot;</span>, model, <span class="st">&quot;</span><span class="sc">\n</span><span class="st">&quot;</span>)</span>
<span id="cb16-8"><a href="#cb16-8" tabindex="-1"></a></span>
<span id="cb16-9"><a href="#cb16-9" tabindex="-1"></a>  <span class="co"># In a real implementation, you would make API calls here</span></span>
<span id="cb16-10"><a href="#cb16-10" tabindex="-1"></a>  <span class="co"># For example:</span></span>
<span id="cb16-11"><a href="#cb16-11" tabindex="-1"></a>  <span class="co"># response &lt;- httr::POST(</span></span>
<span id="cb16-12"><a href="#cb16-12" tabindex="-1"></a>  <span class="co">#   url = &quot;https://api.custom-provider.com/v1/chat/completions&quot;,</span></span>
<span id="cb16-13"><a href="#cb16-13" tabindex="-1"></a>  <span class="co">#   body = list(prompt = prompt, model = model),</span></span>
<span id="cb16-14"><a href="#cb16-14" tabindex="-1"></a>  <span class="co">#   httr::add_headers(Authorization = paste(&quot;Bearer&quot;, api_key)),</span></span>
<span id="cb16-15"><a href="#cb16-15" tabindex="-1"></a>  <span class="co">#   encode = &quot;json&quot;</span></span>
<span id="cb16-16"><a href="#cb16-16" tabindex="-1"></a>  <span class="co"># )</span></span>
<span id="cb16-17"><a href="#cb16-17" tabindex="-1"></a>  <span class="co"># result &lt;- httr::content(response)$choices[[1]]$text</span></span>
<span id="cb16-18"><a href="#cb16-18" tabindex="-1"></a></span>
<span id="cb16-19"><a href="#cb16-19" tabindex="-1"></a>  <span class="co"># For this example, just return a fixed response</span></span>
<span id="cb16-20"><a href="#cb16-20" tabindex="-1"></a>  result <span class="ot">&lt;-</span> <span class="st">&quot;T cells&quot;</span></span>
<span id="cb16-21"><a href="#cb16-21" tabindex="-1"></a>  <span class="fu">return</span>(result)</span>
<span id="cb16-22"><a href="#cb16-22" tabindex="-1"></a>}</span>
<span id="cb16-23"><a href="#cb16-23" tabindex="-1"></a></span>
<span id="cb16-24"><a href="#cb16-24" tabindex="-1"></a><span class="co"># Register the custom provider</span></span>
<span id="cb16-25"><a href="#cb16-25" tabindex="-1"></a><span class="fu">register_custom_provider</span>(</span>
<span id="cb16-26"><a href="#cb16-26" tabindex="-1"></a>  <span class="at">provider_name =</span> <span class="st">&quot;custom_provider&quot;</span>,</span>
<span id="cb16-27"><a href="#cb16-27" tabindex="-1"></a>  <span class="at">process_fn =</span> custom_process_fn,</span>
<span id="cb16-28"><a href="#cb16-28" tabindex="-1"></a>  <span class="at">description =</span> <span class="st">&quot;My custom LLM provider&quot;</span></span>
<span id="cb16-29"><a href="#cb16-29" tabindex="-1"></a>)</span>
<span id="cb16-30"><a href="#cb16-30" tabindex="-1"></a></span>
<span id="cb16-31"><a href="#cb16-31" tabindex="-1"></a><span class="co"># Register a custom model</span></span>
<span id="cb16-32"><a href="#cb16-32" tabindex="-1"></a><span class="fu">register_custom_model</span>(</span>
<span id="cb16-33"><a href="#cb16-33" tabindex="-1"></a>  <span class="at">model_name =</span> <span class="st">&quot;custom-model&quot;</span>,</span>
<span id="cb16-34"><a href="#cb16-34" tabindex="-1"></a>  <span class="at">provider_name =</span> <span class="st">&quot;custom_provider&quot;</span>,</span>
<span id="cb16-35"><a href="#cb16-35" tabindex="-1"></a>  <span class="at">model_config =</span> <span class="fu">list</span>(</span>
<span id="cb16-36"><a href="#cb16-36" tabindex="-1"></a>    <span class="at">temperature =</span> <span class="fl">0.7</span>,</span>
<span id="cb16-37"><a href="#cb16-37" tabindex="-1"></a>    <span class="at">max_tokens =</span> <span class="dv">2000</span></span>
<span id="cb16-38"><a href="#cb16-38" tabindex="-1"></a>  )</span>
<span id="cb16-39"><a href="#cb16-39" tabindex="-1"></a>)</span>
<span id="cb16-40"><a href="#cb16-40" tabindex="-1"></a></span>
<span id="cb16-41"><a href="#cb16-41" tabindex="-1"></a><span class="co"># Use the custom model</span></span>
<span id="cb16-42"><a href="#cb16-42" tabindex="-1"></a><span class="co"># custom_results &lt;- annotate_cell_types(</span></span>
<span id="cb16-43"><a href="#cb16-43" tabindex="-1"></a><span class="co">#   input = marker_data,</span></span>
<span id="cb16-44"><a href="#cb16-44" tabindex="-1"></a><span class="co">#   tissue_name = &quot;human PBMC&quot;,</span></span>
<span id="cb16-45"><a href="#cb16-45" tabindex="-1"></a><span class="co">#   model = &quot;custom-model&quot;,</span></span>
<span id="cb16-46"><a href="#cb16-46" tabindex="-1"></a><span class="co">#   api_key = &quot;your-custom-api-key&quot;</span></span>
<span id="cb16-47"><a href="#cb16-47" tabindex="-1"></a><span class="co"># )</span></span></code></pre></div>
</div>
<div id="using-the-unified-logging-system" class="section level3">
<h3>Using the Unified Logging System</h3>
<p>mLLMCelltype provides a comprehensive unified logging system with
structured output, performance monitoring, and multi-level logging:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="co"># Configure the global logger (recommended approach)</span></span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a><span class="fu">configure_logger</span>(<span class="at">level =</span> <span class="st">&quot;INFO&quot;</span>, <span class="at">console_output =</span> <span class="cn">TRUE</span>, <span class="at">json_format =</span> <span class="cn">TRUE</span>)</span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a></span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a><span class="co"># Use simple logging functions</span></span>
<span id="cb17-5"><a href="#cb17-5" tabindex="-1"></a><span class="fu">log_info</span>(<span class="st">&quot;Starting analysis of cluster 0&quot;</span>, <span class="fu">list</span>(</span>
<span id="cb17-6"><a href="#cb17-6" tabindex="-1"></a>  <span class="at">cluster_id =</span> <span class="st">&quot;0&quot;</span>,</span>
<span id="cb17-7"><a href="#cb17-7" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb17-8"><a href="#cb17-8" tabindex="-1"></a>  <span class="at">marker_genes =</span> <span class="fu">c</span>(<span class="st">&quot;CD3D&quot;</span>, <span class="st">&quot;CD3E&quot;</span>, <span class="st">&quot;CD2&quot;</span>, <span class="st">&quot;IL7R&quot;</span>, <span class="st">&quot;LTB&quot;</span>)</span>
<span id="cb17-9"><a href="#cb17-9" tabindex="-1"></a>))</span>
<span id="cb17-10"><a href="#cb17-10" tabindex="-1"></a></span>
<span id="cb17-11"><a href="#cb17-11" tabindex="-1"></a><span class="co"># Log API calls with performance tracking</span></span>
<span id="cb17-12"><a href="#cb17-12" tabindex="-1"></a><span class="fu">log_info</span>(<span class="st">&quot;API call completed&quot;</span>, <span class="fu">list</span>(</span>
<span id="cb17-13"><a href="#cb17-13" tabindex="-1"></a>  <span class="at">provider =</span> <span class="st">&quot;anthropic&quot;</span>,</span>
<span id="cb17-14"><a href="#cb17-14" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;claude-3.5-sonnet&quot;</span>,</span>
<span id="cb17-15"><a href="#cb17-15" tabindex="-1"></a>  <span class="at">duration_seconds =</span> <span class="fl">2.34</span>,</span>
<span id="cb17-16"><a href="#cb17-16" tabindex="-1"></a>  <span class="at">success =</span> <span class="cn">TRUE</span></span>
<span id="cb17-17"><a href="#cb17-17" tabindex="-1"></a>))</span>
<span id="cb17-18"><a href="#cb17-18" tabindex="-1"></a></span>
<span id="cb17-19"><a href="#cb17-19" tabindex="-1"></a><span class="co"># Log warnings and errors</span></span>
<span id="cb17-20"><a href="#cb17-20" tabindex="-1"></a><span class="fu">log_warn</span>(<span class="st">&quot;Model response had unusual format&quot;</span>, <span class="fu">list</span>(</span>
<span id="cb17-21"><a href="#cb17-21" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;gpt-5&quot;</span>,</span>
<span id="cb17-22"><a href="#cb17-22" tabindex="-1"></a>  <span class="at">response_length =</span> <span class="dv">50</span></span>
<span id="cb17-23"><a href="#cb17-23" tabindex="-1"></a>))</span>
<span id="cb17-24"><a href="#cb17-24" tabindex="-1"></a></span>
<span id="cb17-25"><a href="#cb17-25" tabindex="-1"></a><span class="fu">log_error</span>(<span class="st">&quot;API call failed&quot;</span>, <span class="fu">list</span>(</span>
<span id="cb17-26"><a href="#cb17-26" tabindex="-1"></a>  <span class="at">provider =</span> <span class="st">&quot;openai&quot;</span>,</span>
<span id="cb17-27"><a href="#cb17-27" tabindex="-1"></a>  <span class="at">error =</span> <span class="st">&quot;Rate limit exceeded&quot;</span></span>
<span id="cb17-28"><a href="#cb17-28" tabindex="-1"></a>))</span>
<span id="cb17-29"><a href="#cb17-29" tabindex="-1"></a></span>
<span id="cb17-30"><a href="#cb17-30" tabindex="-1"></a><span class="co"># Alternatively, create a custom logger instance</span></span>
<span id="cb17-31"><a href="#cb17-31" tabindex="-1"></a>custom_logger <span class="ot">&lt;-</span> UnifiedLogger<span class="sc">$</span><span class="fu">new</span>(</span>
<span id="cb17-32"><a href="#cb17-32" tabindex="-1"></a>  <span class="at">base_dir =</span> <span class="st">&quot;custom_logs&quot;</span>,</span>
<span id="cb17-33"><a href="#cb17-33" tabindex="-1"></a>  <span class="at">level =</span> <span class="st">&quot;DEBUG&quot;</span>,</span>
<span id="cb17-34"><a href="#cb17-34" tabindex="-1"></a>  <span class="at">console_output =</span> <span class="cn">TRUE</span>,</span>
<span id="cb17-35"><a href="#cb17-35" tabindex="-1"></a>  <span class="at">json_format =</span> <span class="cn">TRUE</span></span>
<span id="cb17-36"><a href="#cb17-36" tabindex="-1"></a>)</span>
<span id="cb17-37"><a href="#cb17-37" tabindex="-1"></a></span>
<span id="cb17-38"><a href="#cb17-38" tabindex="-1"></a><span class="co"># Use the custom logger</span></span>
<span id="cb17-39"><a href="#cb17-39" tabindex="-1"></a>custom_logger<span class="sc">$</span><span class="fu">info</span>(<span class="st">&quot;Custom log message&quot;</span>, <span class="fu">list</span>(<span class="at">analysis_step =</span> <span class="st">&quot;preprocessing&quot;</span>))</span>
<span id="cb17-40"><a href="#cb17-40" tabindex="-1"></a>custom_logger<span class="sc">$</span><span class="fu">debug</span>(<span class="st">&quot;Detailed debugging info&quot;</span>, <span class="fu">list</span>(<span class="at">variable_state =</span> <span class="st">&quot;initialized&quot;</span>))</span>
<span id="cb17-41"><a href="#cb17-41" tabindex="-1"></a></span>
<span id="cb17-42"><a href="#cb17-42" tabindex="-1"></a><span class="co"># Get performance summary</span></span>
<span id="cb17-43"><a href="#cb17-43" tabindex="-1"></a>performance <span class="ot">&lt;-</span> <span class="fu">get_logger</span>()<span class="sc">$</span><span class="fu">get_performance_summary</span>()</span>
<span id="cb17-44"><a href="#cb17-44" tabindex="-1"></a><span class="fu">print</span>(performance)</span></code></pre></div>
</div>
<div id="using-the-cachemanager" class="section level3">
<h3>Using the CacheManager</h3>
<p>The <code>CacheManager</code> class helps optimize performance by
caching consensus results:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="co"># Create a cache manager</span></span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a>cache_manager <span class="ot">&lt;-</span> CacheManager<span class="sc">$</span><span class="fu">new</span>(<span class="at">cache_dir =</span> <span class="cn">NULL</span>)</span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a></span>
<span id="cb18-4"><a href="#cb18-4" tabindex="-1"></a><span class="co"># Generate a cache key</span></span>
<span id="cb18-5"><a href="#cb18-5" tabindex="-1"></a>cache_key <span class="ot">&lt;-</span> cache_manager<span class="sc">$</span><span class="fu">generate_key</span>(</span>
<span id="cb18-6"><a href="#cb18-6" tabindex="-1"></a>  <span class="at">input =</span> marker_data,</span>
<span id="cb18-7"><a href="#cb18-7" tabindex="-1"></a>  <span class="at">models =</span> <span class="fu">c</span>(<span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>, <span class="st">&quot;gpt-5&quot;</span>),</span>
<span id="cb18-8"><a href="#cb18-8" tabindex="-1"></a>  <span class="at">cluster_id =</span> <span class="st">&quot;0&quot;</span></span>
<span id="cb18-9"><a href="#cb18-9" tabindex="-1"></a>)</span>
<span id="cb18-10"><a href="#cb18-10" tabindex="-1"></a></span>
<span id="cb18-11"><a href="#cb18-11" tabindex="-1"></a><span class="co"># Check if results exist in cache</span></span>
<span id="cb18-12"><a href="#cb18-12" tabindex="-1"></a><span class="cf">if</span> (cache_manager<span class="sc">$</span><span class="fu">has_cache</span>(cache_key)) {</span>
<span id="cb18-13"><a href="#cb18-13" tabindex="-1"></a>  <span class="co"># Load from cache</span></span>
<span id="cb18-14"><a href="#cb18-14" tabindex="-1"></a>  cached_results <span class="ot">&lt;-</span> cache_manager<span class="sc">$</span><span class="fu">load_from_cache</span>(cache_key)</span>
<span id="cb18-15"><a href="#cb18-15" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb18-16"><a href="#cb18-16" tabindex="-1"></a>  <span class="co"># Process and save to cache</span></span>
<span id="cb18-17"><a href="#cb18-17" tabindex="-1"></a>  <span class="co"># results &lt;- process_cluster(...)</span></span>
<span id="cb18-18"><a href="#cb18-18" tabindex="-1"></a>  <span class="co"># cache_manager$save_to_cache(cache_key, results)</span></span>
<span id="cb18-19"><a href="#cb18-19" tabindex="-1"></a>}</span>
<span id="cb18-20"><a href="#cb18-20" tabindex="-1"></a></span>
<span id="cb18-21"><a href="#cb18-21" tabindex="-1"></a><span class="co"># Get cache statistics</span></span>
<span id="cb18-22"><a href="#cb18-22" tabindex="-1"></a>cache_stats <span class="ot">&lt;-</span> cache_manager<span class="sc">$</span><span class="fu">get_cache_stats</span>()</span>
<span id="cb18-23"><a href="#cb18-23" tabindex="-1"></a></span>
<span id="cb18-24"><a href="#cb18-24" tabindex="-1"></a><span class="co"># Clear cache (with confirmation)</span></span>
<span id="cb18-25"><a href="#cb18-25" tabindex="-1"></a><span class="co"># cache_manager$clear_cache(confirm = TRUE)</span></span></code></pre></div>
</div>
<div id="cache-management" class="section level3">
<h3>Cache Management</h3>
<p>mLLMCelltype provides convenient functions for managing cache
directories:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="co"># Check cache location</span></span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a><span class="fu">mllmcelltype_cache_dir</span>()</span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a></span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a><span class="co"># Use local cache</span></span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a><span class="fu">mllmcelltype_cache_dir</span>(<span class="st">&quot;local&quot;</span>)</span>
<span id="cb19-6"><a href="#cb19-6" tabindex="-1"></a></span>
<span id="cb19-7"><a href="#cb19-7" tabindex="-1"></a><span class="co"># Clear cache</span></span>
<span id="cb19-8"><a href="#cb19-8" tabindex="-1"></a><span class="fu">mllmcelltype_clear_cache</span>()</span></code></pre></div>
</div>
</div>
<div id="next-steps" class="section level2">
<h2>Next Steps</h2>
<p>Now that you’ve explored the advanced features of mLLMCelltype, you
can:</p>
<ul>
<li><a href="https://cafferyang.com/mLLMCelltype/articles/contributing-guide.html">Contribute
to the project</a>: Learn how to contribute to mLLMCelltype</li>
<li><a href="https://cafferyang.com/mLLMCelltype/news/index.html">Review
the version history</a>: Explore the development history of the
package</li>
<li><a href="https://cafferyang.com/mLLMCelltype/articles/introduction.html">Return
to the introduction</a>: Review the basic concepts</li>
</ul>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
