<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Chen Yang" />

<meta name="date" content="2026-02-07" />

<title>Getting Started with mLLMCelltype</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Getting Started with mLLMCelltype</h1>
<h4 class="author">Chen Yang</h4>
<h4 class="date">2026-02-07</h4>



<div id="getting-started-with-mllmcelltype" class="section level1">
<h1>Getting Started with mLLMCelltype</h1>
<p>This guide provides a quick introduction to using mLLMCelltype for
cell type annotation in single-cell RNA sequencing data. We’ll cover the
basic workflow, input data requirements, and a simple example to get you
started.</p>
<div id="basic-workflow" class="section level2">
<h2>Basic Workflow</h2>
<p>The mLLMCelltype workflow consists of these main steps:</p>
<ol style="list-style-type: decimal">
<li><strong>Prepare marker gene data</strong> for each cluster</li>
<li><strong>Run annotation</strong> using one or multiple LLMs</li>
<li><strong>Create consensus</strong> from multiple model predictions
(optional)</li>
<li><strong>Integrate results</strong> with your Seurat or Scanpy
object</li>
<li><strong>Visualize results</strong> with uncertainty metrics</li>
</ol>
</div>
<div id="loading-the-package-and-setting-up-api-keys" class="section level2">
<h2>Loading the Package and Setting Up API Keys</h2>
<p>First, load the mLLMCelltype package:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(mLLMCelltype)</span></code></pre></div>
<div id="setting-up-api-keys" class="section level3">
<h3>Setting Up API Keys</h3>
<p>Before using mLLMCelltype, you need to set up API keys for the LLM
providers you plan to use:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># Set API keys as environment variables</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">ANTHROPIC_API_KEY =</span> <span class="st">&quot;your-anthropic-api-key&quot;</span>)  <span class="co"># For Claude models</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">OPENAI_API_KEY =</span> <span class="st">&quot;your-openai-api-key&quot;</span>)        <span class="co"># For GPT models</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">GEMINI_API_KEY =</span> <span class="st">&quot;your-gemini-api-key&quot;</span>)        <span class="co"># For Gemini models</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">OPENROUTER_API_KEY =</span> <span class="st">&quot;your-openrouter-api-key&quot;</span>) <span class="co"># For OpenRouter models</span></span></code></pre></div>
<p>You can obtain API keys from: - Anthropic: <a href="https://console.anthropic.com/" class="uri">https://console.anthropic.com/</a> - OpenAI: <a href="https://platform.openai.com/" class="uri">https://platform.openai.com/</a> - Google (Gemini): <a href="https://ai.google.dev/" class="uri">https://ai.google.dev/</a> -
OpenRouter: <a href="https://openrouter.ai/keys" class="uri">https://openrouter.ai/keys</a></p>
<p>Alternatively, you can provide API keys directly in function
calls:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>  <span class="at">input =</span> markers,</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a>  <span class="at">api_key =</span> <span class="st">&quot;your-anthropic-api-key&quot;</span>,  <span class="co"># Direct API key</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>  <span class="at">top_gene_count =</span> <span class="dv">10</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a>)</span></code></pre></div>
</div>
</div>
<div id="input-data-requirements" class="section level2">
<h2>Input Data Requirements</h2>
<p>mLLMCelltype accepts marker gene data in several formats:</p>
<div id="data-frame-format" class="section level3">
<h3>1. Data Frame Format</h3>
<p>A data frame with the following columns: - <code>cluster</code>:
Cluster ID (preserved as-is from your data) - <code>gene</code>: Gene
name/symbol - <code>avg_log2FC</code> or similar metric: Log fold change
- <code>p_val_adj</code> or similar metric: Adjusted p-value</p>
<p>Example:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># Example marker data frame</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>markers_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>  <span class="at">cluster =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>),</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>  <span class="at">gene =</span> <span class="fu">c</span>(<span class="st">&quot;CD3D&quot;</span>, <span class="st">&quot;CD3E&quot;</span>, <span class="st">&quot;CD2&quot;</span>, <span class="st">&quot;CD14&quot;</span>, <span class="st">&quot;LYZ&quot;</span>, <span class="st">&quot;CST3&quot;</span>),</span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>  <span class="at">avg_log2FC =</span> <span class="fu">c</span>(<span class="fl">2.5</span>, <span class="fl">2.3</span>, <span class="fl">2.1</span>, <span class="fl">3.1</span>, <span class="fl">2.8</span>, <span class="fl">2.5</span>),</span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>  <span class="at">p_val_adj =</span> <span class="fu">c</span>(<span class="fl">0.001</span>, <span class="fl">0.001</span>, <span class="fl">0.002</span>, <span class="fl">0.0001</span>, <span class="fl">0.0002</span>, <span class="fl">0.0005</span>)</span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="seurat-findmarkers-output" class="section level3">
<h3>2. Seurat FindMarkers Output</h3>
<p>You can directly use the output from Seurat’s
<code>FindAllMarkers()</code> function:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># Assuming you have a Seurat object named &#39;seurat_obj&#39;</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>all_markers <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(seurat_obj, <span class="at">only.pos =</span> <span class="cn">TRUE</span>, <span class="at">min.pct =</span> <span class="fl">0.25</span>, <span class="at">logfc.threshold =</span> <span class="fl">0.25</span>)</span></code></pre></div>
</div>
<div id="csv-file-path" class="section level3">
<h3>3. CSV File Path</h3>
<p>A path to a CSV file containing marker gene data:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># Path to your CSV file</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>markers_file <span class="ot">&lt;-</span> <span class="st">&quot;path/to/markers.csv&quot;</span></span></code></pre></div>
</div>
<div id="list-format" class="section level3">
<h3>4. List Format</h3>
<p>A list where each element contains marker genes for a cluster:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="co"># Example marker list</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>markers_list <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>  <span class="st">&quot;0&quot;</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;CD3D&quot;</span>, <span class="st">&quot;CD3E&quot;</span>, <span class="st">&quot;CD2&quot;</span>, <span class="st">&quot;IL7R&quot;</span>, <span class="st">&quot;LTB&quot;</span>),</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>  <span class="st">&quot;1&quot;</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&quot;CD14&quot;</span>, <span class="st">&quot;LYZ&quot;</span>, <span class="st">&quot;CST3&quot;</span>, <span class="st">&quot;MS4A7&quot;</span>, <span class="st">&quot;FCGR3A&quot;</span>)</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>)</span></code></pre></div>
</div>
</div>
<div id="function-parameters" class="section level2">
<h2>Function Parameters</h2>
<p>The <code>annotate_cell_types</code> function has the following
parameters:</p>
<table>
<colgroup>
<col width="28%" />
<col width="33%" />
<col width="38%" />
</colgroup>
<thead>
<tr class="header">
<th>Parameter</th>
<th>Description</th>
<th>Default Value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><code>input</code></td>
<td>Marker gene data (data frame, list, or file path)</td>
<td>(required)</td>
</tr>
<tr class="even">
<td><code>tissue_name</code></td>
<td>Tissue name (e.g., “human PBMC”, “mouse brain”)</td>
<td><code>NULL</code></td>
</tr>
<tr class="odd">
<td><code>model</code></td>
<td>LLM model to use</td>
<td><code>&quot;gpt-5&quot;</code></td>
</tr>
<tr class="even">
<td><code>api_key</code></td>
<td>API key (if not set in environment)</td>
<td><code>NA</code></td>
</tr>
<tr class="odd">
<td><code>top_gene_count</code></td>
<td>Number of top genes per cluster to use</td>
<td><code>10</code></td>
</tr>
<tr class="even">
<td><code>debug</code></td>
<td>Whether to print debugging information</td>
<td><code>FALSE</code></td>
</tr>
</tbody>
</table>
<p>Note: If <code>api_key</code> is set to <code>NA</code>, the function
will return the generated prompt without making an API call, which is
useful for reviewing the prompt before sending it to the API.</p>
</div>
<div id="basic-usage-example" class="section level2">
<h2>Basic Usage Example</h2>
<p>Here’s a simple example using a single LLM model for annotation:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co"># Example marker data</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>markers <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>  <span class="at">cluster =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>, <span class="dv">1</span>),</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>  <span class="at">gene =</span> <span class="fu">c</span>(<span class="st">&quot;CD3D&quot;</span>, <span class="st">&quot;CD3E&quot;</span>, <span class="st">&quot;CD2&quot;</span>, <span class="st">&quot;IL7R&quot;</span>, <span class="st">&quot;LTB&quot;</span>, <span class="st">&quot;CD14&quot;</span>, <span class="st">&quot;LYZ&quot;</span>, <span class="st">&quot;CST3&quot;</span>, <span class="st">&quot;MS4A7&quot;</span>, <span class="st">&quot;FCGR3A&quot;</span>),</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>  <span class="at">avg_log2FC =</span> <span class="fu">c</span>(<span class="fl">2.5</span>, <span class="fl">2.3</span>, <span class="fl">2.1</span>, <span class="fl">1.8</span>, <span class="fl">1.7</span>, <span class="fl">3.1</span>, <span class="fl">2.8</span>, <span class="fl">2.5</span>, <span class="fl">2.2</span>, <span class="fl">2.0</span>),</span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a>  <span class="at">p_val_adj =</span> <span class="fu">c</span>(<span class="fl">0.001</span>, <span class="fl">0.001</span>, <span class="fl">0.002</span>, <span class="fl">0.003</span>, <span class="fl">0.005</span>, <span class="fl">0.0001</span>, <span class="fl">0.0002</span>, <span class="fl">0.0005</span>, <span class="fl">0.001</span>, <span class="fl">0.002</span>)</span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a>)</span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a></span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a><span class="co"># Run annotation with a single model</span></span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a>  <span class="at">input =</span> markers,</span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,</span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a>  <span class="at">api_key =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>),</span>
<span id="cb8-15"><a href="#cb8-15" tabindex="-1"></a>  <span class="at">top_gene_count =</span> <span class="dv">10</span>,</span>
<span id="cb8-16"><a href="#cb8-16" tabindex="-1"></a>  <span class="at">debug =</span> <span class="cn">FALSE</span>  <span class="co"># Set to TRUE for more detailed output</span></span>
<span id="cb8-17"><a href="#cb8-17" tabindex="-1"></a>)</span>
<span id="cb8-18"><a href="#cb8-18" tabindex="-1"></a></span>
<span id="cb8-19"><a href="#cb8-19" tabindex="-1"></a><span class="co"># Print results</span></span>
<span id="cb8-20"><a href="#cb8-20" tabindex="-1"></a><span class="fu">print</span>(results)</span></code></pre></div>
<div id="example-output" class="section level3">
<h3>Example Output</h3>
<p>When using a single model like Claude, the output will be a character
vector with one annotation per cluster:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="sc">&gt;</span> <span class="fu">print</span>(results)</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>[<span class="dv">1</span>] <span class="st">&quot;0: T cells&quot;</span>   <span class="st">&quot;1: Monocytes&quot;</span></span></code></pre></div>
</div>
</div>
<div id="multi-model-consensus-example" class="section level2">
<h2>Multi-Model Consensus Example</h2>
<p>For more reliable annotations, you can use multiple models and create
a consensus:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="co"># Define models to use</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>models <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>  <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,  <span class="co"># Anthropic</span></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a>  <span class="st">&quot;gpt-5&quot;</span>,                      <span class="co"># OpenAI</span></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a>  <span class="st">&quot;gemini-1.5-pro&quot;</span>               <span class="co"># Google</span></span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a>)</span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a></span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a><span class="co"># API keys for different providers</span></span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a>api_keys <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a>  <span class="at">anthropic =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>),</span>
<span id="cb10-11"><a href="#cb10-11" tabindex="-1"></a>  <span class="at">openai =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;OPENAI_API_KEY&quot;</span>),</span>
<span id="cb10-12"><a href="#cb10-12" tabindex="-1"></a>  <span class="at">gemini =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;GEMINI_API_KEY&quot;</span>)</span>
<span id="cb10-13"><a href="#cb10-13" tabindex="-1"></a>)</span>
<span id="cb10-14"><a href="#cb10-14" tabindex="-1"></a></span>
<span id="cb10-15"><a href="#cb10-15" tabindex="-1"></a><span class="co"># Run annotation with multiple models</span></span>
<span id="cb10-16"><a href="#cb10-16" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb10-17"><a href="#cb10-17" tabindex="-1"></a><span class="cf">for</span> (model <span class="cf">in</span> models) {</span>
<span id="cb10-18"><a href="#cb10-18" tabindex="-1"></a>  provider <span class="ot">&lt;-</span> <span class="fu">get_provider</span>(model)</span>
<span id="cb10-19"><a href="#cb10-19" tabindex="-1"></a>  api_key <span class="ot">&lt;-</span> api_keys[[provider]]</span>
<span id="cb10-20"><a href="#cb10-20" tabindex="-1"></a></span>
<span id="cb10-21"><a href="#cb10-21" tabindex="-1"></a>  results[[model]] <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb10-22"><a href="#cb10-22" tabindex="-1"></a>    <span class="at">input =</span> markers,</span>
<span id="cb10-23"><a href="#cb10-23" tabindex="-1"></a>    <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb10-24"><a href="#cb10-24" tabindex="-1"></a>    <span class="at">model =</span> model,</span>
<span id="cb10-25"><a href="#cb10-25" tabindex="-1"></a>    <span class="at">api_key =</span> api_key,</span>
<span id="cb10-26"><a href="#cb10-26" tabindex="-1"></a>    <span class="at">top_gene_count =</span> <span class="dv">10</span></span>
<span id="cb10-27"><a href="#cb10-27" tabindex="-1"></a>  )</span>
<span id="cb10-28"><a href="#cb10-28" tabindex="-1"></a>}</span>
<span id="cb10-29"><a href="#cb10-29" tabindex="-1"></a></span>
<span id="cb10-30"><a href="#cb10-30" tabindex="-1"></a><span class="co"># Create consensus</span></span>
<span id="cb10-31"><a href="#cb10-31" tabindex="-1"></a>consensus_results <span class="ot">&lt;-</span> <span class="fu">interactive_consensus_annotation</span>(</span>
<span id="cb10-32"><a href="#cb10-32" tabindex="-1"></a>  <span class="at">input =</span> markers,</span>
<span id="cb10-33"><a href="#cb10-33" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb10-34"><a href="#cb10-34" tabindex="-1"></a>  <span class="at">models =</span> models,  <span class="co"># Use all the models defined above</span></span>
<span id="cb10-35"><a href="#cb10-35" tabindex="-1"></a>  <span class="at">api_keys =</span> api_keys,</span>
<span id="cb10-36"><a href="#cb10-36" tabindex="-1"></a>  <span class="at">controversy_threshold =</span> <span class="fl">0.7</span>,</span>
<span id="cb10-37"><a href="#cb10-37" tabindex="-1"></a>  <span class="at">entropy_threshold =</span> <span class="fl">1.0</span>,</span>
<span id="cb10-38"><a href="#cb10-38" tabindex="-1"></a>  <span class="at">consensus_check_model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span></span>
<span id="cb10-39"><a href="#cb10-39" tabindex="-1"></a>)</span></code></pre></div>
<div id="consensus-output-example" class="section level3">
<h3>Consensus Output Example</h3>
<p>The function automatically prints a summary upon completion:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="sc">&gt;</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>Consensus Summary<span class="sc">:</span></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="sc">-----------------</span></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>Total clusters<span class="sc">:</span> <span class="dv">2</span></span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>Controversial clusters<span class="sc">:</span> <span class="dv">0</span></span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a>Consensus achieved <span class="cf">for</span> all clusters</span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a></span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a>Cluster <span class="dv">0</span><span class="sc">:</span></span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a>  Final annotation<span class="sc">:</span> T cells</span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a>  Consensus proportion<span class="sc">:</span> <span class="fl">1.0</span></span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a>  Entropy<span class="sc">:</span> <span class="fl">0.0</span></span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a>  Model predictions<span class="sc">:</span></span>
<span id="cb11-13"><a href="#cb11-13" tabindex="-1"></a>    <span class="sc">-</span> claude<span class="sc">-</span>sonnet<span class="dv">-4-5-20250929</span><span class="sc">:</span> T cells</span>
<span id="cb11-14"><a href="#cb11-14" tabindex="-1"></a>    <span class="sc">-</span> gpt<span class="dv">-5</span><span class="sc">:</span> T cells</span>
<span id="cb11-15"><a href="#cb11-15" tabindex="-1"></a>    <span class="sc">-</span> gemini<span class="fl">-2.5</span><span class="sc">-</span>pro<span class="sc">:</span> T cells</span>
<span id="cb11-16"><a href="#cb11-16" tabindex="-1"></a></span>
<span id="cb11-17"><a href="#cb11-17" tabindex="-1"></a>Cluster <span class="dv">1</span><span class="sc">:</span></span>
<span id="cb11-18"><a href="#cb11-18" tabindex="-1"></a>  Final annotation<span class="sc">:</span> Monocytes</span>
<span id="cb11-19"><a href="#cb11-19" tabindex="-1"></a>  Consensus proportion<span class="sc">:</span> <span class="fl">1.0</span></span>
<span id="cb11-20"><a href="#cb11-20" tabindex="-1"></a>  Entropy<span class="sc">:</span> <span class="fl">0.0</span></span>
<span id="cb11-21"><a href="#cb11-21" tabindex="-1"></a>  Model predictions<span class="sc">:</span></span>
<span id="cb11-22"><a href="#cb11-22" tabindex="-1"></a>    <span class="sc">-</span> claude<span class="sc">-</span>sonnet<span class="dv">-4-5-20250929</span><span class="sc">:</span> Monocytes</span>
<span id="cb11-23"><a href="#cb11-23" tabindex="-1"></a>    <span class="sc">-</span> gpt<span class="dv">-5</span><span class="sc">:</span> Monocytes</span>
<span id="cb11-24"><a href="#cb11-24" tabindex="-1"></a>    <span class="sc">-</span> gemini<span class="fl">-2.5</span><span class="sc">-</span>pro<span class="sc">:</span> Monocytes</span></code></pre></div>
</div>
</div>
<div id="integrating-with-seurat" class="section level2">
<h2>Integrating with Seurat</h2>
<p>To add the annotations to your Seurat object:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="co"># Assuming you have a Seurat object named &#39;seurat_obj&#39; and consensus results</span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a><span class="co"># Add consensus annotations to Seurat object</span></span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a>seurat_obj<span class="sc">$</span>cell_type_consensus <span class="ot">&lt;-</span> plyr<span class="sc">::</span><span class="fu">mapvalues</span>(</span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">as.character</span>(<span class="fu">Idents</span>(seurat_obj)),</span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a>  <span class="at">from =</span> <span class="fu">names</span>(consensus_results<span class="sc">$</span>final_annotations),</span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a>  <span class="at">to =</span> consensus_results<span class="sc">$</span>final_annotations</span>
<span id="cb12-9"><a href="#cb12-9" tabindex="-1"></a>)</span>
<span id="cb12-10"><a href="#cb12-10" tabindex="-1"></a></span>
<span id="cb12-11"><a href="#cb12-11" tabindex="-1"></a><span class="co"># Extract consensus metrics from the consensus results</span></span>
<span id="cb12-12"><a href="#cb12-12" tabindex="-1"></a><span class="co"># Note: These metrics are available in the consensus_results$initial_results$consensus_results</span></span>
<span id="cb12-13"><a href="#cb12-13" tabindex="-1"></a>consensus_metrics <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">names</span>(consensus_results<span class="sc">$</span>initial_results<span class="sc">$</span>consensus_results), <span class="cf">function</span>(cluster_id) {</span>
<span id="cb12-14"><a href="#cb12-14" tabindex="-1"></a>  metrics <span class="ot">&lt;-</span> consensus_results<span class="sc">$</span>initial_results<span class="sc">$</span>consensus_results[[cluster_id]]</span>
<span id="cb12-15"><a href="#cb12-15" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">list</span>(</span>
<span id="cb12-16"><a href="#cb12-16" tabindex="-1"></a>    <span class="at">cluster =</span> cluster_id,</span>
<span id="cb12-17"><a href="#cb12-17" tabindex="-1"></a>    <span class="at">consensus_proportion =</span> metrics<span class="sc">$</span>consensus_proportion,</span>
<span id="cb12-18"><a href="#cb12-18" tabindex="-1"></a>    <span class="at">entropy =</span> metrics<span class="sc">$</span>entropy</span>
<span id="cb12-19"><a href="#cb12-19" tabindex="-1"></a>  ))</span>
<span id="cb12-20"><a href="#cb12-20" tabindex="-1"></a>})</span>
<span id="cb12-21"><a href="#cb12-21" tabindex="-1"></a></span>
<span id="cb12-22"><a href="#cb12-22" tabindex="-1"></a><span class="co"># Convert to data frame for easier handling</span></span>
<span id="cb12-23"><a href="#cb12-23" tabindex="-1"></a>metrics_df <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, <span class="fu">lapply</span>(consensus_metrics, data.frame))</span>
<span id="cb12-24"><a href="#cb12-24" tabindex="-1"></a></span>
<span id="cb12-25"><a href="#cb12-25" tabindex="-1"></a><span class="co"># Add consensus proportion to Seurat object</span></span>
<span id="cb12-26"><a href="#cb12-26" tabindex="-1"></a>seurat_obj<span class="sc">$</span>consensus_proportion <span class="ot">&lt;-</span> plyr<span class="sc">::</span><span class="fu">mapvalues</span>(</span>
<span id="cb12-27"><a href="#cb12-27" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">as.character</span>(<span class="fu">Idents</span>(seurat_obj)),</span>
<span id="cb12-28"><a href="#cb12-28" tabindex="-1"></a>  <span class="at">from =</span> metrics_df<span class="sc">$</span>cluster,</span>
<span id="cb12-29"><a href="#cb12-29" tabindex="-1"></a>  <span class="at">to =</span> metrics_df<span class="sc">$</span>consensus_proportion</span>
<span id="cb12-30"><a href="#cb12-30" tabindex="-1"></a>)</span>
<span id="cb12-31"><a href="#cb12-31" tabindex="-1"></a></span>
<span id="cb12-32"><a href="#cb12-32" tabindex="-1"></a><span class="co"># Add entropy to Seurat object</span></span>
<span id="cb12-33"><a href="#cb12-33" tabindex="-1"></a>seurat_obj<span class="sc">$</span>entropy <span class="ot">&lt;-</span> plyr<span class="sc">::</span><span class="fu">mapvalues</span>(</span>
<span id="cb12-34"><a href="#cb12-34" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">as.character</span>(<span class="fu">Idents</span>(seurat_obj)),</span>
<span id="cb12-35"><a href="#cb12-35" tabindex="-1"></a>  <span class="at">from =</span> metrics_df<span class="sc">$</span>cluster,</span>
<span id="cb12-36"><a href="#cb12-36" tabindex="-1"></a>  <span class="at">to =</span> metrics_df<span class="sc">$</span>entropy</span>
<span id="cb12-37"><a href="#cb12-37" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="basic-visualization" class="section level2">
<h2>Basic Visualization</h2>
<p>Here’s a simple visualization of the results using Seurat:</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="co"># Plot UMAP with cell type annotations</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a><span class="fu">DimPlot</span>(seurat_obj, <span class="at">group.by =</span> <span class="st">&quot;cell_type_consensus&quot;</span>, <span class="at">label =</span> <span class="cn">TRUE</span>, <span class="at">repel =</span> <span class="cn">TRUE</span>) <span class="sc">+</span></span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Cell Type Annotations&quot;</span>) <span class="sc">+</span></span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>))</span></code></pre></div>
</div>
<div id="understanding-the-output" class="section level2">
<h2>Understanding the Output</h2>
<p>The output of <code>annotate_cell_types()</code> is a vector of cell
type annotations, where each element corresponds to a cluster.</p>
<p>The output of <code>interactive_consensus_annotation()</code> is a
list containing:</p>
<ul>
<li><code>final_annotations</code>: Final consensus cell type
annotations</li>
<li><code>initial_results</code>: Initial predictions from each
model</li>
<li><code>controversial_clusters</code>: List of clusters that required
discussion</li>
<li><code>discussion_logs</code>: Detailed logs of the discussion
process</li>
<li><code>session_id</code>: Unique identifier for the annotation
session</li>
</ul>
<div id="understanding-uncertainty-metrics" class="section level3">
<h3>Understanding Uncertainty Metrics</h3>
<p>When using consensus annotation, two key metrics help evaluate the
reliability of annotations:</p>
<ul>
<li><strong>Consensus Proportion</strong>: Ranges from 0 to 1,
indicating the proportion of models that agree on the final annotation.
Higher values indicate stronger agreement.</li>
<li><strong>Entropy</strong>: Measures the uncertainty in model
predictions. Lower values indicate more certainty. An entropy of 0 means
all models agree perfectly.</li>
</ul>
<p>Clusters with low consensus proportion or high entropy may require
manual review.</p>
</div>
</div>
<div id="using-openrouter-free-models" class="section level2">
<h2>Using OpenRouter Free Models</h2>
<p>If you don’t have access to paid API keys, you can use OpenRouter’s
free models:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="co"># Set OpenRouter API key</span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">OPENROUTER_API_KEY =</span> <span class="st">&quot;your-openrouter-api-key&quot;</span>)</span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a></span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a><span class="co"># Use a free model</span></span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a>free_results <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a>  <span class="at">input =</span> markers,</span>
<span id="cb14-7"><a href="#cb14-7" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb14-8"><a href="#cb14-8" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;meta-llama/llama-4-maverick:free&quot;</span>,  <span class="co"># Note the :free suffix</span></span>
<span id="cb14-9"><a href="#cb14-9" tabindex="-1"></a>  <span class="at">api_key =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;OPENROUTER_API_KEY&quot;</span>),</span>
<span id="cb14-10"><a href="#cb14-10" tabindex="-1"></a>  <span class="at">top_gene_count =</span> <span class="dv">10</span></span>
<span id="cb14-11"><a href="#cb14-11" tabindex="-1"></a>)</span>
<span id="cb14-12"><a href="#cb14-12" tabindex="-1"></a></span>
<span id="cb14-13"><a href="#cb14-13" tabindex="-1"></a><span class="co"># Print results</span></span>
<span id="cb14-14"><a href="#cb14-14" tabindex="-1"></a><span class="fu">print</span>(free_results)</span></code></pre></div>
<p>Available free models (Updated Oct 2025):</p>
<ul>
<li><code>meta-llama/llama-4-maverick:free</code> - Meta Llama 4
Maverick (256K context, best performance)</li>
<li><code>deepseek/deepseek-r1:free</code> - DeepSeek R1 (advanced
reasoning)</li>
<li><code>meta-llama/llama-3.3-70b-instruct:free</code> - Meta Llama 3.3
70B (reliable)</li>
<li><code>venice/uncensored:free</code> - Venice Uncensored (new
model)</li>
<li><code>minimax/minimax-m2:free</code> - MiniMax M2 (optimized for
coding)</li>
<li><code>z-ai/glm-4.5-air:free</code> - GLM 4.5 Air (lightweight)</li>
</ul>
<p><strong>Important</strong>: OpenRouter reduced free tier limits in
2025: - <strong>Free accounts</strong>: 50 requests/day (down from 200),
20 requests/minute - <strong>Accounts with $10+ credits</strong>: 1000
requests/day - <strong>Some models removed</strong>: NVIDIA Nemotron and
others have exited the free tier - <strong>For production use</strong>:
Consider using paid models for better reliability</p>
</div>
<div id="troubleshooting" class="section level2">
<h2>Troubleshooting</h2>
<div id="common-issues" class="section level3">
<h3>Common Issues</h3>
<ol style="list-style-type: decimal">
<li><p><strong>API Key Not Found</strong>:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a>Error<span class="sc">:</span> No auth credentials found</span></code></pre></div>
<p><strong>Solution</strong>: Ensure you’ve set the correct API key
environment variable or provided it directly in the function
call.</p></li>
<li><p><strong>Rate Limiting</strong>:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a>Error<span class="sc">:</span> Rate limit exceeded</span></code></pre></div>
<p><strong>Solution</strong>: Wait a few minutes before trying again, or
reduce the number of API calls by processing fewer clusters at
once.</p></li>
<li><p><strong>Invalid Model Name</strong>:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a>Error<span class="sc">:</span> Unsupported model<span class="sc">:</span> [model_name]</span></code></pre></div>
<p><strong>Solution</strong>: Check that you’re using a supported model
name and that it’s spelled correctly.</p></li>
<li><p><strong>Network Issues</strong>:</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a>Error<span class="sc">:</span> Could not connect to API</span></code></pre></div>
<p><strong>Solution</strong>: Check your internet connection and try
again. If the problem persists, the API service might be down.</p></li>
</ol>
</div>
</div>
<div id="next-steps" class="section level2">
<h2>Next Steps</h2>
<p>Now that you understand the basics of mLLMCelltype, you can
explore:</p>
<ul>
<li><a href="https://cafferyang.com/mLLMCelltype/articles/usage-tutorial.html">Usage
Tutorial</a>: More detailed usage examples</li>
<li><a href="https://cafferyang.com/mLLMCelltype/articles/consensus-principles.html">Consensus
Annotation Principles</a>: Learn about the consensus mechanism</li>
<li><a href="https://cafferyang.com/mLLMCelltype/articles/visualization-guide.html">Visualization
Guide</a>: Create publication-ready visualizations</li>
<li><a href="https://cafferyang.com/mLLMCelltype/articles/advanced-features.html">Advanced
Features</a>: Explore hierarchical annotation and other advanced
features</li>
<li><a href="https://cafferyang.com/mLLMCelltype/articles/faq.html">FAQ</a>:
Answers to common questions</li>
</ul>
<p>If you encounter any issues, check the <a href="https://cafferyang.com/mLLMCelltype/articles/faq.html">FAQ</a> or
<a href="https://github.com/cafferychen777/mLLMCelltype/issues">open an
issue</a> on our GitHub repository.</p>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
