<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>mLLMCelltype: Overview and Quick Reference</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">mLLMCelltype: Overview and Quick
Reference</h1>



<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(mLLMCelltype)</span></code></pre></div>
<div id="overview" class="section level2">
<h2>Overview</h2>
<p>mLLMCelltype is an R package that leverages various large language
models (LLMs) for automated cell type annotation in single-cell RNA
sequencing data. It implements a consensus-based approach where multiple
LLMs collaborate to provide more reliable annotations. For more details,
please refer to our paper (<a href="doi:10.1101/2025.04.10.647852" class="uri">doi:10.1101/2025.04.10.647852</a>).</p>
</div>
<div id="installation" class="section level2">
<h2>Installation</h2>
<p>You can install the released version of mLLMCelltype from CRAN:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;mLLMCelltype&quot;</span>)</span></code></pre></div>
<p>Or install the development version from GitHub:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># install.packages(&quot;devtools&quot;)</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;cafferychen777/mLLMCelltype&quot;</span>)</span></code></pre></div>
</div>
<div id="setting-up-api-keys" class="section level2">
<h2>Setting Up API Keys</h2>
<p>Before using mLLMCelltype, you need to set up API keys for the LLMs
you want to use. The package supports multiple LLM providers including
OpenAI, Anthropic, Google (Gemini), and X.AI (Grok).</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># Set API keys (recommended to use .Renviron or .env file)</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="co"># OpenAI API key</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">OPENAI_API_KEY =</span> <span class="st">&quot;your-openai-api-key&quot;</span>)</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="co"># Anthropic API key</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">ANTHROPIC_API_KEY =</span> <span class="st">&quot;your-anthropic-api-key&quot;</span>)</span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a><span class="co"># Google API key</span></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">GOOGLE_API_KEY =</span> <span class="st">&quot;your-google-api-key&quot;</span>)</span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a></span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a><span class="co"># X.AI API key</span></span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">XAI_API_KEY =</span> <span class="st">&quot;your-xai-api-key&quot;</span>)</span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a></span>
<span id="cb4-14"><a href="#cb4-14" tabindex="-1"></a><span class="co"># Alternatively, use .env file</span></span>
<span id="cb4-15"><a href="#cb4-15" tabindex="-1"></a><span class="co"># dotenv::load_dot_env()</span></span></code></pre></div>
</div>
<div id="basic-usage" class="section level2">
<h2>Basic Usage</h2>
<div id="annotating-cell-types-with-seurat-object" class="section level3">
<h3>Annotating Cell Types with Seurat Object</h3>
<p>The main function of mLLMCelltype is to annotate cell types in a
Seurat object:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># Load required packages</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="fu">library</span>(mLLMCelltype)</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a><span class="co"># Load your preprocessed Seurat object</span></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="co"># pbmc &lt;- readRDS(&quot;your_seurat_object.rds&quot;)</span></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="co"># Find marker genes for each cluster</span></span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a><span class="co"># pbmc_markers &lt;- FindAllMarkers(pbmc,</span></span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a><span class="co">#                           only.pos = TRUE,</span></span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a><span class="co">#                           min.pct = 0.25,</span></span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a><span class="co">#                           logfc.threshold = 0.25)</span></span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a></span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a><span class="co"># Cache will be automatically managed using system cache directory</span></span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a></span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a><span class="co"># Run LLMCelltype annotation with multiple LLM models</span></span>
<span id="cb5-18"><a href="#cb5-18" tabindex="-1"></a>consensus_results <span class="ot">&lt;-</span> <span class="fu">interactive_consensus_annotation</span>(</span>
<span id="cb5-19"><a href="#cb5-19" tabindex="-1"></a>  <span class="at">input =</span> pbmc_markers,</span>
<span id="cb5-20"><a href="#cb5-20" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,  <span class="co"># provide tissue context</span></span>
<span id="cb5-21"><a href="#cb5-21" tabindex="-1"></a>  <span class="at">models =</span> <span class="fu">c</span>(</span>
<span id="cb5-22"><a href="#cb5-22" tabindex="-1"></a>    <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,  <span class="co"># Anthropic</span></span>
<span id="cb5-23"><a href="#cb5-23" tabindex="-1"></a>    <span class="st">&quot;gpt-5&quot;</span>,                   <span class="co"># OpenAI</span></span>
<span id="cb5-24"><a href="#cb5-24" tabindex="-1"></a>    <span class="st">&quot;gemini-2.5-pro&quot;</span>            <span class="co"># Google</span></span>
<span id="cb5-25"><a href="#cb5-25" tabindex="-1"></a>  ),</span>
<span id="cb5-26"><a href="#cb5-26" tabindex="-1"></a>  <span class="at">api_keys =</span> <span class="fu">list</span>(</span>
<span id="cb5-27"><a href="#cb5-27" tabindex="-1"></a>    <span class="at">anthropic =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>),</span>
<span id="cb5-28"><a href="#cb5-28" tabindex="-1"></a>    <span class="at">openai =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;OPENAI_API_KEY&quot;</span>),</span>
<span id="cb5-29"><a href="#cb5-29" tabindex="-1"></a>    <span class="at">gemini =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;GOOGLE_API_KEY&quot;</span>)</span>
<span id="cb5-30"><a href="#cb5-30" tabindex="-1"></a>  ),</span>
<span id="cb5-31"><a href="#cb5-31" tabindex="-1"></a>  <span class="at">top_gene_count =</span> <span class="dv">10</span>,</span>
<span id="cb5-32"><a href="#cb5-32" tabindex="-1"></a>  <span class="at">controversy_threshold =</span> <span class="fl">1.0</span>,</span>
<span id="cb5-33"><a href="#cb5-33" tabindex="-1"></a>  <span class="at">entropy_threshold =</span> <span class="fl">1.0</span>,</span>
<span id="cb5-34"><a href="#cb5-34" tabindex="-1"></a>  <span class="at">consensus_check_model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,  <span class="co"># Use a reliable model for consensus checking</span></span>
<span id="cb5-35"><a href="#cb5-35" tabindex="-1"></a>  <span class="at">cache_dir =</span> <span class="cn">NULL</span>  <span class="co"># Uses default system cache directory</span></span>
<span id="cb5-36"><a href="#cb5-36" tabindex="-1"></a>)</span>
<span id="cb5-37"><a href="#cb5-37" tabindex="-1"></a></span>
<span id="cb5-38"><a href="#cb5-38" tabindex="-1"></a><span class="co"># Add annotations to Seurat object</span></span>
<span id="cb5-39"><a href="#cb5-39" tabindex="-1"></a><span class="co"># Get cell type annotations from consensus_results$final_annotations</span></span>
<span id="cb5-40"><a href="#cb5-40" tabindex="-1"></a>cluster_to_celltype_map <span class="ot">&lt;-</span> consensus_results<span class="sc">$</span>final_annotations</span>
<span id="cb5-41"><a href="#cb5-41" tabindex="-1"></a></span>
<span id="cb5-42"><a href="#cb5-42" tabindex="-1"></a><span class="co"># Create new cell type identifier column</span></span>
<span id="cb5-43"><a href="#cb5-43" tabindex="-1"></a>cell_types <span class="ot">&lt;-</span> <span class="fu">as.character</span>(<span class="fu">Idents</span>(pbmc))</span>
<span id="cb5-44"><a href="#cb5-44" tabindex="-1"></a><span class="cf">for</span> (cluster_id <span class="cf">in</span> <span class="fu">names</span>(cluster_to_celltype_map)) {</span>
<span id="cb5-45"><a href="#cb5-45" tabindex="-1"></a>  cell_types[cell_types <span class="sc">==</span> cluster_id] <span class="ot">&lt;-</span> cluster_to_celltype_map[[cluster_id]]</span>
<span id="cb5-46"><a href="#cb5-46" tabindex="-1"></a>}</span>
<span id="cb5-47"><a href="#cb5-47" tabindex="-1"></a></span>
<span id="cb5-48"><a href="#cb5-48" tabindex="-1"></a><span class="co"># Add cell types to Seurat object</span></span>
<span id="cb5-49"><a href="#cb5-49" tabindex="-1"></a>pbmc<span class="sc">$</span>mLLM_cell_type <span class="ot">&lt;-</span> cell_types</span></code></pre></div>
</div>
<div id="visualizing-results" class="section level3">
<h3>Visualizing Results</h3>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># Plot UMAP with cell type annotations</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="fu">library</span>(cowplot)</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a></span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a><span class="co"># Basic visualization</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(pbmc, <span class="at">group.by =</span> <span class="st">&quot;mLLM_cell_type&quot;</span>, <span class="at">label =</span> <span class="cn">TRUE</span>) <span class="sc">+</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;mLLMCelltype Consensus Annotations&quot;</span>)</span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a></span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a><span class="co"># Visualize consensus proportion (confidence metric)</span></span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a>pbmc<span class="sc">$</span>consensus_proportion <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a><span class="cf">for</span> (cluster_id <span class="cf">in</span> <span class="fu">names</span>(consensus_results<span class="sc">$</span>initial_results<span class="sc">$</span>consensus_results)) {</span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a>  cluster_cells <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">Idents</span>(pbmc) <span class="sc">==</span> cluster_id)</span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a>  pbmc<span class="sc">$</span>consensus_proportion[cluster_cells] <span class="ot">&lt;-</span></span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a>    consensus_results<span class="sc">$</span>initial_results<span class="sc">$</span>consensus_results[[cluster_id]]<span class="sc">$</span>consensus_proportion</span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a>}</span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a></span>
<span id="cb6-18"><a href="#cb6-18" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">FeaturePlot</span>(pbmc, <span class="at">features =</span> <span class="st">&quot;consensus_proportion&quot;</span>,</span>
<span id="cb6-19"><a href="#cb6-19" tabindex="-1"></a>                 <span class="at">min.cutoff =</span> <span class="dv">0</span>, <span class="at">max.cutoff =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb6-20"><a href="#cb6-20" tabindex="-1"></a>  <span class="fu">scale_color_gradientn</span>(<span class="at">colors =</span> <span class="fu">c</span>(<span class="st">&quot;blue&quot;</span>, <span class="st">&quot;green&quot;</span>, <span class="st">&quot;red&quot;</span>)) <span class="sc">+</span></span>
<span id="cb6-21"><a href="#cb6-21" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Consensus Proportion&quot;</span>) <span class="sc">+</span></span>
<span id="cb6-22"><a href="#cb6-22" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">&quot;right&quot;</span>)</span></code></pre></div>
</div>
</div>
<div id="supported-models" class="section level2">
<h2>Supported Models</h2>
<p>mLLMCelltype supports various LLM models:</p>
<ul>
<li><strong>OpenAI</strong>: gpt-5, gpt-5-mini, gpt-5-nano, gpt-4.1,
gpt-4.1-mini, gpt-4.1-nano, gpt-4o, gpt-4o-mini, gpt-4-turbo,
gpt-3.5-turbo, o1, o1-mini, o1-preview, o1-pro</li>
<li><strong>Anthropic</strong>: claude-sonnet-4-5-20250929,
claude-opus-4-1-20250805, claude-sonnet-4-20250514,
claude-opus-4-20250514, claude-haiku-4-20250514,
claude-3-5-haiku-20241022, claude-3-opus-20240229</li>
<li><strong>DeepSeek</strong>: deepseek-chat, deepseek-reasoner</li>
<li><strong>Google</strong>: gemini-2.5-pro, gemini-2.0-flash,
gemini-2.0-flash-exp, gemini-1.5-pro, gemini-1.5-flash</li>
<li><strong>Qwen</strong>: qwen-max-2025-01-25, qwen3-72b</li>
<li><strong>Stepfun</strong>: step-2-mini, step-2-16k, step-1-8k</li>
<li><strong>Zhipu</strong>: glm-4-plus, glm-3-turbo</li>
<li><strong>MiniMax</strong>: minimax-text-01</li>
<li><strong>Grok</strong>: grok-3, grok-3-latest, grok-3-fast,
grok-3-fast-latest, grok-3-mini, grok-3-mini-latest, grok-3-mini-fast,
grok-3-mini-fast-latest</li>
<li><strong>OpenRouter</strong>: Access to models from multiple
providers through a single API</li>
</ul>
</div>
<div id="advanced-usage" class="section level2">
<h2>Advanced Usage</h2>
<div id="using-a-single-llm-model" class="section level3">
<h3>Using a Single LLM Model</h3>
<p>If you only want to use a single LLM model instead of the consensus
approach:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="co"># Run annotation with a single model</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>single_model_results <span class="ot">&lt;-</span> <span class="fu">annotate_with_single_model</span>(</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>  <span class="at">input =</span> pbmc_markers,</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,</span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a>  <span class="at">api_key =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>),</span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a>  <span class="at">top_gene_count =</span> <span class="dv">10</span>,</span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a>  <span class="at">cache_dir =</span> <span class="cn">NULL</span>  <span class="co"># Uses default system cache directory</span></span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a>)</span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a></span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a><span class="co"># Add annotations to Seurat object</span></span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a>pbmc<span class="sc">$</span>single_model_cell_type <span class="ot">&lt;-</span> plyr<span class="sc">::</span><span class="fu">mapvalues</span>(</span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">as.character</span>(<span class="fu">Idents</span>(pbmc)),</span>
<span id="cb7-14"><a href="#cb7-14" tabindex="-1"></a>  <span class="at">from =</span> <span class="fu">names</span>(single_model_results),</span>
<span id="cb7-15"><a href="#cb7-15" tabindex="-1"></a>  <span class="at">to =</span> single_model_results</span>
<span id="cb7-16"><a href="#cb7-16" tabindex="-1"></a>)</span>
<span id="cb7-17"><a href="#cb7-17" tabindex="-1"></a></span>
<span id="cb7-18"><a href="#cb7-18" tabindex="-1"></a><span class="co"># Visualize results</span></span>
<span id="cb7-19"><a href="#cb7-19" tabindex="-1"></a><span class="fu">DimPlot</span>(pbmc, <span class="at">group.by =</span> <span class="st">&quot;single_model_cell_type&quot;</span>, <span class="at">label =</span> <span class="cn">TRUE</span>) <span class="sc">+</span></span>
<span id="cb7-20"><a href="#cb7-20" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Cell Types Annotated by Single LLM Model&quot;</span>)</span></code></pre></div>
</div>
<div id="customizing-consensus-parameters" class="section level3">
<h3>Customizing Consensus Parameters</h3>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co"># Customize consensus parameters</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>custom_consensus_results <span class="ot">&lt;-</span> <span class="fu">interactive_consensus_annotation</span>(</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>  <span class="at">input =</span> pbmc_markers,</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>  <span class="at">models =</span> <span class="fu">c</span>(<span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>, <span class="st">&quot;gpt-5&quot;</span>, <span class="st">&quot;gemini-2.5-pro&quot;</span>),</span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a>  <span class="at">api_keys =</span> <span class="fu">list</span>(</span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a>    <span class="at">anthropic =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>),</span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a>    <span class="at">openai =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;OPENAI_API_KEY&quot;</span>),</span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a>    <span class="at">gemini =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;GOOGLE_API_KEY&quot;</span>)</span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a>  ),</span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a>  <span class="at">top_gene_count =</span> <span class="dv">15</span>,                <span class="co"># Number of top marker genes to use</span></span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a>  <span class="at">controversy_threshold =</span> <span class="fl">0.7</span>,        <span class="co"># Threshold for controversy detection</span></span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a>  <span class="at">entropy_threshold =</span> <span class="fl">0.7</span>,            <span class="co"># Entropy threshold for controversy detection</span></span>
<span id="cb8-14"><a href="#cb8-14" tabindex="-1"></a>  <span class="at">max_discussion_rounds =</span> <span class="dv">2</span>,          <span class="co"># Maximum rounds of discussion</span></span>
<span id="cb8-15"><a href="#cb8-15" tabindex="-1"></a>  <span class="at">consensus_check_model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,  <span class="co"># Use a reliable model for consensus checking</span></span>
<span id="cb8-16"><a href="#cb8-16" tabindex="-1"></a>  <span class="at">cache_dir =</span> <span class="cn">NULL</span>  <span class="co"># Uses default system cache directory</span></span>
<span id="cb8-17"><a href="#cb8-17" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="using-custom-providers" class="section level3">
<h3>Using Custom Providers</h3>
<p>You can register custom LLM providers:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="co"># Register a custom provider</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="fu">register_custom_provider</span>(</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>  <span class="at">provider_name =</span> <span class="st">&quot;my_custom_provider&quot;</span>,</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>  <span class="at">process_function =</span> my_process_function,</span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>  <span class="at">models =</span> <span class="fu">c</span>(<span class="st">&quot;my-model-1&quot;</span>, <span class="st">&quot;my-model-2&quot;</span>)</span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a>)</span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a><span class="co"># Use the custom provider</span></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a>  <span class="at">seurat_obj =</span> seurat_obj,</span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a>  <span class="at">models =</span> <span class="fu">c</span>(<span class="st">&quot;my-model-1&quot;</span>, <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>),</span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a>  <span class="co"># other parameters</span></span>
<span id="cb9-13"><a href="#cb9-13" tabindex="-1"></a>)</span></code></pre></div>
</div>
</div>
<div id="caching-results" class="section level2">
<h2>Caching Results</h2>
<p>mLLMCelltype includes a caching system to save API costs:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="co"># Enable caching</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">MLLM_CACHE_ENABLED =</span> <span class="st">&quot;TRUE&quot;</span>)</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a><span class="co"># Set cache directory</span></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">MLLM_CACHE_DIR =</span> <span class="st">&quot;path/to/cache&quot;</span>)</span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a></span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a><span class="co"># Clear cache if needed</span></span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a><span class="fu">clear_cache</span>()</span></code></pre></div>
</div>
<div id="conclusion" class="section level2">
<h2>Conclusion</h2>
<p>mLLMCelltype is a framework for automated cell type annotation using
large language models. By combining predictions from multiple models, it
aims to produce more reliable annotations than single-model
approaches.</p>
<p>For bug reports or feature requests, please visit: <a href="https://github.com/cafferychen777/mLLMCelltype/issues">https://github.com/cafferychen777/mLLMCelltype/issues</a>.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
