<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Chen Yang" />

<meta name="date" content="2026-02-07" />

<title>Usage Tutorial</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Usage Tutorial</h1>
<h4 class="author">Chen Yang</h4>
<h4 class="date">2026-02-07</h4>



<div id="usage-tutorial" class="section level1">
<h1>Usage Tutorial</h1>
<p>This tutorial provides detailed instructions for using mLLMCelltype
for cell type annotation in single-cell RNA sequencing data. We’ll cover
various usage scenarios, parameter configurations, and integration with
Seurat.</p>
<div id="comprehensive-function-parameters" class="section level2">
<h2>Comprehensive Function Parameters</h2>
<div id="annotate_cell_types" class="section level3">
<h3>annotate_cell_types()</h3>
<p>The main function for cell type annotation with a single model:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(mLLMCelltype)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a>  input,                <span class="co"># Marker gene data (data frame, list, or file path)</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a>  tissue_name,          <span class="co"># Tissue name (e.g., &quot;human PBMC&quot;, &quot;mouse brain&quot;)</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a>  model,                <span class="co"># LLM model to use</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a>  <span class="at">api_key =</span> <span class="cn">NA</span>,         <span class="co"># API key (if not set in environment, NA returns prompt only)</span></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a>  <span class="at">top_gene_count =</span> <span class="dv">10</span>,  <span class="co"># Number of top genes per cluster to use</span></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a>  <span class="at">debug =</span> <span class="cn">FALSE</span>         <span class="co"># Whether to print debugging information</span></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="interactive_consensus_annotation" class="section level3">
<h3>interactive_consensus_annotation()</h3>
<p>Function for creating consensus annotations from multiple models
through interactive discussion:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>consensus_results <span class="ot">&lt;-</span> <span class="fu">interactive_consensus_annotation</span>(</span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>  input,                <span class="co"># Original marker gene data (Seurat FindAllMarkers result or list of genes)</span></span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="cn">NULL</span>,   <span class="co"># Optional tissue name</span></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a>  <span class="at">models =</span> <span class="fu">c</span>(<span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>, <span class="st">&quot;gpt-5&quot;</span>, <span class="st">&quot;gemini-1.5-pro&quot;</span>),  <span class="co"># Models to use</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a>  api_keys,             <span class="co"># Named list of API keys</span></span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a>  <span class="at">top_gene_count =</span> <span class="dv">10</span>,  <span class="co"># Number of top genes to use</span></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a>  <span class="at">controversy_threshold =</span> <span class="fl">0.7</span>,  <span class="co"># Threshold for identifying controversial clusters</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a>  <span class="at">entropy_threshold =</span> <span class="fl">1.0</span>,  <span class="co"># Entropy threshold for controversial clusters</span></span>
<span id="cb2-9"><a href="#cb2-9" tabindex="-1"></a>  <span class="at">max_discussion_rounds =</span> <span class="dv">3</span>,  <span class="co"># Maximum discussion rounds</span></span>
<span id="cb2-10"><a href="#cb2-10" tabindex="-1"></a>  <span class="at">consensus_check_model =</span> <span class="cn">NULL</span>,  <span class="co"># Model to use for consensus checking (see recommendations below)</span></span>
<span id="cb2-11"><a href="#cb2-11" tabindex="-1"></a>  <span class="at">log_dir =</span> <span class="st">&quot;logs&quot;</span>,     <span class="co"># Directory for logs</span></span>
<span id="cb2-12"><a href="#cb2-12" tabindex="-1"></a>  <span class="at">cache_dir =</span> <span class="cn">NULL</span>,  <span class="co"># Uses default system cache directory</span></span>
<span id="cb2-13"><a href="#cb2-13" tabindex="-1"></a>  <span class="at">use_cache =</span> <span class="cn">TRUE</span>      <span class="co"># Whether to use cache</span></span>
<span id="cb2-14"><a href="#cb2-14" tabindex="-1"></a>)</span></code></pre></div>
<p><strong>Important Note on
<code>consensus_check_model</code></strong>: This parameter is used for
evaluating semantic similarity, calculating consensus metrics, and
moderating discussions. We recommend using capable models such as: -
<code>claude-sonnet-4-5-20250929</code> (Anthropic) -
<code>claude-opus-4-1-20250805</code> (Anthropic) - <code>o1</code>
(OpenAI) - <code>gpt-5</code> (OpenAI) - <code>gemini-2.5-pro</code>
(Google)</p>
<p>See the main README for detailed recommendations and examples.</p>
</div>
</div>
<div id="detailed-usage-scenarios" class="section level2">
<h2>Detailed Usage Scenarios</h2>
<div id="scenario-1-basic-annotation-with-a-single-model" class="section level3">
<h3>Scenario 1: Basic Annotation with a Single Model</h3>
<p>For quick exploration or when API usage is a concern:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># Load example data</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;pbmc_small&quot;</span>)</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="co"># Find markers</span></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a>pbmc_markers <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(pbmc_small, <span class="at">only.pos =</span> <span class="cn">TRUE</span>, <span class="at">min.pct =</span> <span class="fl">0.25</span>, <span class="at">logfc.threshold =</span> <span class="fl">0.25</span>)</span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a></span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a><span class="co"># Run annotation with a single model</span></span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>  <span class="at">input =</span> pbmc_markers,</span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,</span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a>  <span class="at">api_key =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>),</span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a>  <span class="at">top_gene_count =</span> <span class="dv">10</span></span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a>)</span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a></span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a><span class="co"># Add annotations to Seurat object</span></span>
<span id="cb3-18"><a href="#cb3-18" tabindex="-1"></a>pbmc_small<span class="sc">$</span>cell_type_claude <span class="ot">&lt;-</span> plyr<span class="sc">::</span><span class="fu">mapvalues</span>(</span>
<span id="cb3-19"><a href="#cb3-19" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">as.character</span>(<span class="fu">Idents</span>(pbmc_small)),</span>
<span id="cb3-20"><a href="#cb3-20" tabindex="-1"></a>  <span class="at">from =</span> <span class="fu">names</span>(results),</span>
<span id="cb3-21"><a href="#cb3-21" tabindex="-1"></a>  <span class="at">to =</span> results</span>
<span id="cb3-22"><a href="#cb3-22" tabindex="-1"></a>)</span>
<span id="cb3-23"><a href="#cb3-23" tabindex="-1"></a></span>
<span id="cb3-24"><a href="#cb3-24" tabindex="-1"></a><span class="co"># Visualize</span></span>
<span id="cb3-25"><a href="#cb3-25" tabindex="-1"></a><span class="fu">DimPlot</span>(pbmc_small, <span class="at">group.by =</span> <span class="st">&quot;cell_type_claude&quot;</span>, <span class="at">label =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
</div>
<div id="scenario-2-multi-model-consensus-for-high-accuracy" class="section level3">
<h3>Scenario 2: Multi-Model Consensus for High Accuracy</h3>
<p>For publication-quality annotations with uncertainty
quantification:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co"># Define multiple models to use</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>models <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a>  <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,  <span class="co"># Anthropic</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>  <span class="st">&quot;gpt-5&quot;</span>,                      <span class="co"># OpenAI</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>  <span class="st">&quot;gemini-1.5-pro&quot;</span>,              <span class="co"># Google</span></span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>  <span class="st">&quot;grok-3&quot;</span>                       <span class="co"># X.AI</span></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>)</span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a></span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a><span class="co"># API keys for different providers</span></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a>api_keys <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a>  <span class="at">anthropic =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>),</span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a>  <span class="at">openai =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;OPENAI_API_KEY&quot;</span>),</span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a>  <span class="at">gemini =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;GEMINI_API_KEY&quot;</span>),</span>
<span id="cb4-14"><a href="#cb4-14" tabindex="-1"></a>  <span class="at">grok =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;GROK_API_KEY&quot;</span>)</span>
<span id="cb4-15"><a href="#cb4-15" tabindex="-1"></a>)</span>
<span id="cb4-16"><a href="#cb4-16" tabindex="-1"></a></span>
<span id="cb4-17"><a href="#cb4-17" tabindex="-1"></a><span class="co"># Run annotation with multiple models</span></span>
<span id="cb4-18"><a href="#cb4-18" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb4-19"><a href="#cb4-19" tabindex="-1"></a><span class="cf">for</span> (model <span class="cf">in</span> models) {</span>
<span id="cb4-20"><a href="#cb4-20" tabindex="-1"></a>  provider <span class="ot">&lt;-</span> <span class="fu">get_provider</span>(model)</span>
<span id="cb4-21"><a href="#cb4-21" tabindex="-1"></a>  api_key <span class="ot">&lt;-</span> api_keys[[provider]]</span>
<span id="cb4-22"><a href="#cb4-22" tabindex="-1"></a></span>
<span id="cb4-23"><a href="#cb4-23" tabindex="-1"></a>  results[[model]] <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb4-24"><a href="#cb4-24" tabindex="-1"></a>    <span class="at">input =</span> pbmc_markers,</span>
<span id="cb4-25"><a href="#cb4-25" tabindex="-1"></a>    <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb4-26"><a href="#cb4-26" tabindex="-1"></a>    <span class="at">model =</span> model,</span>
<span id="cb4-27"><a href="#cb4-27" tabindex="-1"></a>    <span class="at">api_key =</span> api_key,</span>
<span id="cb4-28"><a href="#cb4-28" tabindex="-1"></a>    <span class="at">top_gene_count =</span> <span class="dv">10</span></span>
<span id="cb4-29"><a href="#cb4-29" tabindex="-1"></a>  )</span>
<span id="cb4-30"><a href="#cb4-30" tabindex="-1"></a>}</span>
<span id="cb4-31"><a href="#cb4-31" tabindex="-1"></a></span>
<span id="cb4-32"><a href="#cb4-32" tabindex="-1"></a><span class="co"># Create consensus</span></span>
<span id="cb4-33"><a href="#cb4-33" tabindex="-1"></a>consensus_results <span class="ot">&lt;-</span> <span class="fu">interactive_consensus_annotation</span>(</span>
<span id="cb4-34"><a href="#cb4-34" tabindex="-1"></a>  <span class="at">input =</span> pbmc_markers,</span>
<span id="cb4-35"><a href="#cb4-35" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb4-36"><a href="#cb4-36" tabindex="-1"></a>  <span class="at">models =</span> models,  <span class="co"># Use all the models defined above</span></span>
<span id="cb4-37"><a href="#cb4-37" tabindex="-1"></a>  <span class="at">api_keys =</span> api_keys,</span>
<span id="cb4-38"><a href="#cb4-38" tabindex="-1"></a>  <span class="at">controversy_threshold =</span> <span class="fl">0.7</span>,</span>
<span id="cb4-39"><a href="#cb4-39" tabindex="-1"></a>  <span class="at">entropy_threshold =</span> <span class="fl">1.0</span>,</span>
<span id="cb4-40"><a href="#cb4-40" tabindex="-1"></a>  <span class="at">consensus_check_model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span></span>
<span id="cb4-41"><a href="#cb4-41" tabindex="-1"></a>)</span>
<span id="cb4-42"><a href="#cb4-42" tabindex="-1"></a></span>
<span id="cb4-43"><a href="#cb4-43" tabindex="-1"></a><span class="co"># View consensus results</span></span>
<span id="cb4-44"><a href="#cb4-44" tabindex="-1"></a><span class="co"># You can access the final annotations with consensus_results$final_annotations</span></span>
<span id="cb4-45"><a href="#cb4-45" tabindex="-1"></a></span>
<span id="cb4-46"><a href="#cb4-46" tabindex="-1"></a><span class="co"># Add consensus annotations and metrics to Seurat object</span></span>
<span id="cb4-47"><a href="#cb4-47" tabindex="-1"></a>pbmc_small<span class="sc">$</span>cell_type_consensus <span class="ot">&lt;-</span> plyr<span class="sc">::</span><span class="fu">mapvalues</span>(</span>
<span id="cb4-48"><a href="#cb4-48" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">as.character</span>(<span class="fu">Idents</span>(pbmc_small)),</span>
<span id="cb4-49"><a href="#cb4-49" tabindex="-1"></a>  <span class="at">from =</span> <span class="fu">names</span>(consensus_results<span class="sc">$</span>final_annotations),</span>
<span id="cb4-50"><a href="#cb4-50" tabindex="-1"></a>  <span class="at">to =</span> consensus_results<span class="sc">$</span>final_annotations</span>
<span id="cb4-51"><a href="#cb4-51" tabindex="-1"></a>)</span>
<span id="cb4-52"><a href="#cb4-52" tabindex="-1"></a></span>
<span id="cb4-53"><a href="#cb4-53" tabindex="-1"></a><span class="co"># Extract consensus metrics from the consensus results</span></span>
<span id="cb4-54"><a href="#cb4-54" tabindex="-1"></a>consensus_metrics <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">names</span>(consensus_results<span class="sc">$</span>initial_results<span class="sc">$</span>consensus_results), <span class="cf">function</span>(cluster_id) {</span>
<span id="cb4-55"><a href="#cb4-55" tabindex="-1"></a>  metrics <span class="ot">&lt;-</span> consensus_results<span class="sc">$</span>initial_results<span class="sc">$</span>consensus_results[[cluster_id]]</span>
<span id="cb4-56"><a href="#cb4-56" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">list</span>(</span>
<span id="cb4-57"><a href="#cb4-57" tabindex="-1"></a>    <span class="at">cluster =</span> cluster_id,</span>
<span id="cb4-58"><a href="#cb4-58" tabindex="-1"></a>    <span class="at">consensus_proportion =</span> metrics<span class="sc">$</span>consensus_proportion,</span>
<span id="cb4-59"><a href="#cb4-59" tabindex="-1"></a>    <span class="at">entropy =</span> metrics<span class="sc">$</span>entropy</span>
<span id="cb4-60"><a href="#cb4-60" tabindex="-1"></a>  ))</span>
<span id="cb4-61"><a href="#cb4-61" tabindex="-1"></a>})</span>
<span id="cb4-62"><a href="#cb4-62" tabindex="-1"></a></span>
<span id="cb4-63"><a href="#cb4-63" tabindex="-1"></a><span class="co"># Convert to data frame for easier handling</span></span>
<span id="cb4-64"><a href="#cb4-64" tabindex="-1"></a>metrics_df <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, <span class="fu">lapply</span>(consensus_metrics, data.frame))</span>
<span id="cb4-65"><a href="#cb4-65" tabindex="-1"></a></span>
<span id="cb4-66"><a href="#cb4-66" tabindex="-1"></a><span class="co"># Add consensus proportion to Seurat object</span></span>
<span id="cb4-67"><a href="#cb4-67" tabindex="-1"></a>pbmc_small<span class="sc">$</span>consensus_proportion <span class="ot">&lt;-</span> plyr<span class="sc">::</span><span class="fu">mapvalues</span>(</span>
<span id="cb4-68"><a href="#cb4-68" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">as.character</span>(<span class="fu">Idents</span>(pbmc_small)),</span>
<span id="cb4-69"><a href="#cb4-69" tabindex="-1"></a>  <span class="at">from =</span> metrics_df<span class="sc">$</span>cluster,</span>
<span id="cb4-70"><a href="#cb4-70" tabindex="-1"></a>  <span class="at">to =</span> metrics_df<span class="sc">$</span>consensus_proportion</span>
<span id="cb4-71"><a href="#cb4-71" tabindex="-1"></a>)</span>
<span id="cb4-72"><a href="#cb4-72" tabindex="-1"></a></span>
<span id="cb4-73"><a href="#cb4-73" tabindex="-1"></a><span class="co"># Add entropy to Seurat object</span></span>
<span id="cb4-74"><a href="#cb4-74" tabindex="-1"></a>pbmc_small<span class="sc">$</span>shannon_entropy <span class="ot">&lt;-</span> plyr<span class="sc">::</span><span class="fu">mapvalues</span>(</span>
<span id="cb4-75"><a href="#cb4-75" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">as.character</span>(<span class="fu">Idents</span>(pbmc_small)),</span>
<span id="cb4-76"><a href="#cb4-76" tabindex="-1"></a>  <span class="at">from =</span> metrics_df<span class="sc">$</span>cluster,</span>
<span id="cb4-77"><a href="#cb4-77" tabindex="-1"></a>  <span class="at">to =</span> metrics_df<span class="sc">$</span>entropy</span>
<span id="cb4-78"><a href="#cb4-78" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="scenario-2b-using-free-openrouter-models" class="section level3">
<h3>Scenario 2b: Using Free OpenRouter Models</h3>
<p>For users with limited API credits or budget constraints:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># Set OpenRouter API key</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>openrouter_api_key <span class="ot">&lt;-</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;OPENROUTER_API_KEY&quot;</span>)</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="co"># Define free OpenRouter models to use</span></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a>free_models <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>  <span class="st">&quot;meta-llama/llama-4-maverick:free&quot;</span>,                <span class="co"># Meta Llama 4 Maverick (free)</span></span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a>  <span class="st">&quot;nvidia/llama-3.1-nemotron-ultra-253b-v1:free&quot;</span>,    <span class="co"># NVIDIA Nemotron Ultra 253B (free)</span></span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a>  <span class="st">&quot;deepseek/deepseek-r1:free&quot;</span>,             <span class="co"># DeepSeek R1 (free, advanced reasoning)</span></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a>  <span class="st">&quot;meta-llama/llama-3.3-70b-instruct:free&quot;</span>          <span class="co"># Meta Llama 3.3 70B (free)</span></span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a>)</span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a></span>
<span id="cb5-12"><a href="#cb5-12" tabindex="-1"></a><span class="co"># Run annotation with free OpenRouter models</span></span>
<span id="cb5-13"><a href="#cb5-13" tabindex="-1"></a>free_results <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb5-14"><a href="#cb5-14" tabindex="-1"></a><span class="cf">for</span> (model <span class="cf">in</span> free_models) {</span>
<span id="cb5-15"><a href="#cb5-15" tabindex="-1"></a>  free_results[[model]] <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb5-16"><a href="#cb5-16" tabindex="-1"></a>    <span class="at">input =</span> pbmc_markers,</span>
<span id="cb5-17"><a href="#cb5-17" tabindex="-1"></a>    <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb5-18"><a href="#cb5-18" tabindex="-1"></a>    <span class="at">model =</span> model,  <span class="co"># OpenRouter models are automatically detected by format: &#39;provider/model-name:free&#39;</span></span>
<span id="cb5-19"><a href="#cb5-19" tabindex="-1"></a>    <span class="at">api_key =</span> openrouter_api_key,</span>
<span id="cb5-20"><a href="#cb5-20" tabindex="-1"></a>    <span class="at">top_gene_count =</span> <span class="dv">10</span></span>
<span id="cb5-21"><a href="#cb5-21" tabindex="-1"></a>  )</span>
<span id="cb5-22"><a href="#cb5-22" tabindex="-1"></a>}</span>
<span id="cb5-23"><a href="#cb5-23" tabindex="-1"></a></span>
<span id="cb5-24"><a href="#cb5-24" tabindex="-1"></a><span class="co"># Create consensus with free models</span></span>
<span id="cb5-25"><a href="#cb5-25" tabindex="-1"></a>free_consensus_results <span class="ot">&lt;-</span> <span class="fu">interactive_consensus_annotation</span>(</span>
<span id="cb5-26"><a href="#cb5-26" tabindex="-1"></a>  <span class="at">input =</span> pbmc_markers,</span>
<span id="cb5-27"><a href="#cb5-27" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb5-28"><a href="#cb5-28" tabindex="-1"></a>  <span class="at">models =</span> free_models,  <span class="co"># Use all the free models defined above</span></span>
<span id="cb5-29"><a href="#cb5-29" tabindex="-1"></a>  <span class="at">api_keys =</span> <span class="fu">list</span>(<span class="st">&quot;openrouter&quot;</span> <span class="ot">=</span> openrouter_api_key),</span>
<span id="cb5-30"><a href="#cb5-30" tabindex="-1"></a>  <span class="at">controversy_threshold =</span> <span class="fl">0.7</span>,</span>
<span id="cb5-31"><a href="#cb5-31" tabindex="-1"></a>  <span class="at">entropy_threshold =</span> <span class="fl">1.0</span>,</span>
<span id="cb5-32"><a href="#cb5-32" tabindex="-1"></a>  <span class="at">consensus_check_model =</span> <span class="st">&quot;meta-llama/llama-4-maverick:free&quot;</span>  <span class="co"># Use a free model for consensus checking</span></span>
<span id="cb5-33"><a href="#cb5-33" tabindex="-1"></a>)</span>
<span id="cb5-34"><a href="#cb5-34" tabindex="-1"></a></span>
<span id="cb5-35"><a href="#cb5-35" tabindex="-1"></a><span class="co"># View free model consensus results</span></span>
<span id="cb5-36"><a href="#cb5-36" tabindex="-1"></a><span class="co"># You can access the final annotations with free_consensus_results$final_annotations</span></span>
<span id="cb5-37"><a href="#cb5-37" tabindex="-1"></a></span>
<span id="cb5-38"><a href="#cb5-38" tabindex="-1"></a><span class="co"># Add free model consensus annotations to Seurat object</span></span>
<span id="cb5-39"><a href="#cb5-39" tabindex="-1"></a>pbmc_small<span class="sc">$</span>free_model_consensus <span class="ot">&lt;-</span> plyr<span class="sc">::</span><span class="fu">mapvalues</span>(</span>
<span id="cb5-40"><a href="#cb5-40" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">as.character</span>(<span class="fu">Idents</span>(pbmc_small)),</span>
<span id="cb5-41"><a href="#cb5-41" tabindex="-1"></a>  <span class="at">from =</span> <span class="fu">names</span>(free_consensus_results<span class="sc">$</span>final_annotations),</span>
<span id="cb5-42"><a href="#cb5-42" tabindex="-1"></a>  <span class="at">to =</span> free_consensus_results<span class="sc">$</span>final_annotations</span>
<span id="cb5-43"><a href="#cb5-43" tabindex="-1"></a>)</span>
<span id="cb5-44"><a href="#cb5-44" tabindex="-1"></a></span>
<span id="cb5-45"><a href="#cb5-45" tabindex="-1"></a><span class="co"># Compare paid vs. free model results</span></span>
<span id="cb5-46"><a href="#cb5-46" tabindex="-1"></a>comparison <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb5-47"><a href="#cb5-47" tabindex="-1"></a>  <span class="at">cluster =</span> <span class="fu">names</span>(consensus_results<span class="sc">$</span>final_annotations),</span>
<span id="cb5-48"><a href="#cb5-48" tabindex="-1"></a>  <span class="at">paid_models =</span> consensus_results<span class="sc">$</span>final_annotations,</span>
<span id="cb5-49"><a href="#cb5-49" tabindex="-1"></a>  <span class="at">free_models =</span> free_consensus_results<span class="sc">$</span>final_annotations,</span>
<span id="cb5-50"><a href="#cb5-50" tabindex="-1"></a>  <span class="at">agreement =</span> consensus_results<span class="sc">$</span>final_annotations <span class="sc">==</span> free_consensus_results<span class="sc">$</span>final_annotations</span>
<span id="cb5-51"><a href="#cb5-51" tabindex="-1"></a>)</span>
<span id="cb5-52"><a href="#cb5-52" tabindex="-1"></a><span class="fu">print</span>(comparison)</span></code></pre></div>
</div>
<div id="scenario-3-working-with-csv-files" class="section level3">
<h3>Scenario 3: Working with CSV Files</h3>
<p>For users who prefer working with files:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="co"># Save markers to CSV</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="fu">write.csv</span>(pbmc_markers, <span class="st">&quot;pbmc_markers.csv&quot;</span>, <span class="at">row.names =</span> <span class="cn">FALSE</span>)</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a><span class="co"># Run annotation using the CSV file</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>  <span class="at">input =</span> <span class="st">&quot;pbmc_markers.csv&quot;</span>,</span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,</span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a>  <span class="at">api_key =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>)</span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="scenario-4-custom-caching" class="section level3">
<h3>Scenario 4: Custom Caching</h3>
<p>For better control over caching behavior:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="co"># Note: The annotate_cell_types function does not have built-in caching.</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="co"># If you need caching, you can implement it separately.</span></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a><span class="co"># Run annotation</span></span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a>  <span class="at">input =</span> pbmc_markers,</span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,</span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a>  <span class="at">api_key =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>),</span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a>  <span class="at">top_gene_count =</span> <span class="dv">10</span>,</span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a>  <span class="at">debug =</span> <span class="cn">FALSE</span></span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a>)</span>
<span id="cb7-13"><a href="#cb7-13" tabindex="-1"></a></span>
<span id="cb7-14"><a href="#cb7-14" tabindex="-1"></a><span class="co"># If you need custom caching, you can implement it using your own cache manager</span></span>
<span id="cb7-15"><a href="#cb7-15" tabindex="-1"></a><span class="co"># This is just a conceptual example and not part of the actual package</span></span>
<span id="cb7-16"><a href="#cb7-16" tabindex="-1"></a><span class="co"># cache_manager &lt;- YourCacheManager$new(cache_dir = &quot;path/to/cache&quot;)</span></span>
<span id="cb7-17"><a href="#cb7-17" tabindex="-1"></a><span class="co"># cache_manager$clear_cache()</span></span></code></pre></div>
</div>
</div>
<div id="model-selection-guide" class="section level2">
<h2>Model Selection Guide</h2>
<p>mLLMCelltype supports a wide range of LLM models. Here’s a guide to
help you choose:</p>
<div id="high-performance-models" class="section level3">
<h3>High Performance Models</h3>
<p>For the most accurate annotations:</p>
<ul>
<li><strong>Anthropic Claude Sonnet 4.5</strong>
(<code>claude-sonnet-4-5-20250929</code>)</li>
<li><strong>Anthropic Claude Opus 4.1</strong>
(<code>claude-opus-4-1-20250805</code>)</li>
<li><strong>OpenAI GPT-5</strong> (<code>gpt-5</code>)</li>
<li><strong>Google Gemini 2.5 Pro</strong>
(<code>gemini-2.5-pro</code>)</li>
</ul>
</div>
<div id="balanced-performancecost-models" class="section level3">
<h3>Balanced Performance/Cost Models</h3>
<p>For good results with lower API costs:</p>
<ul>
<li><strong>X.AI Grok-3</strong> (<code>grok-3</code>): Competitive
performance at lower cost</li>
<li><strong>DeepSeek V3</strong> (<code>deepseek-v3</code>): Good
performance for specialized tissues</li>
</ul>
</div>
<div id="economy-models" class="section level3">
<h3>Economy Models</h3>
<p>For preliminary exploration or large datasets:</p>
<ul>
<li><strong>Qwen 2.5</strong> (<code>qwen-max-2025-01-25</code>): Good
performance for the cost</li>
<li><strong>Zhipu GLM-4</strong> (<code>glm-4</code>): Economical option
with decent performance</li>
<li><strong>MiniMax</strong> (<code>minimax</code>): Cost-effective for
initial exploration</li>
</ul>
</div>
<div id="free-models-via-openrouter" class="section level3">
<h3>Free Models via OpenRouter</h3>
<p>For users with limited API credits or budget constraints:</p>
<ul>
<li><strong>Meta Llama 4 Maverick</strong>
(<code>meta-llama/llama-4-maverick:free</code>): Most reliable and fast,
recommended for consensus checking</li>
<li><strong>NVIDIA Nemotron Ultra 253B</strong>
(<code>nvidia/llama-3.1-nemotron-ultra-253b-v1:free</code>): Good
performance with consistent formatting</li>
<li><strong>DeepSeek R1</strong>
(<code>deepseek/deepseek-r1:free</code>): Advanced reasoning model
(free)</li>
</ul>
<p>Based on our testing, we recommend the following free models:</p>
<ul>
<li><code>meta-llama/llama-4-maverick:free</code>: Most reliable and
fast, recommended for consensus checking</li>
<li><code>nvidia/llama-3.1-nemotron-ultra-253b-v1:free</code>: Good
performance with consistent formatting</li>
<li><code>deepseek/deepseek-r1:free</code>: Reliable with good response
time</li>
</ul>
<p>Some models may have limitations:</p>
<ul>
<li><code>deepseek/deepseek-r1:free</code>: May occasionally return
empty results</li>
<li><code>thudm/glm-z1-9b:free</code>: May return localized error
messages (“non-character parameter”) when used for consensus
checking</li>
</ul>
<p>These free models are accessed through OpenRouter and don’t consume
credits, but may have limitations compared to paid models. Use the
<code>:free</code> suffix in the model name to access them.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co"># Example of using a free model via OpenRouter</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="co"># First, set your OpenRouter API key</span></span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a><span class="fu">Sys.setenv</span>(<span class="at">OPENROUTER_API_KEY =</span> <span class="st">&quot;your-openrouter-api-key&quot;</span>)</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a></span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a><span class="co"># Then use a free model with the :free suffix</span></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a>free_model_results <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a>  <span class="at">input =</span> pbmc_markers,</span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;meta-llama/llama-4-maverick:free&quot;</span>,  <span class="co"># Note the :free suffix</span></span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a>  <span class="at">api_key =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;OPENROUTER_API_KEY&quot;</span>)</span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a>  <span class="co"># No need to specify provider - it&#39;s automatically detected from the model name format</span></span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a>)</span></code></pre></div>
</div>
</div>
<div id="integration-with-seurat-workflow" class="section level2">
<h2>Integration with Seurat Workflow</h2>
<p>Here’s a complete example of integrating mLLMCelltype with a Seurat
workflow:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="fu">library</span>(mLLMCelltype)</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a><span class="co"># Load data</span></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a><span class="fu">data</span>(<span class="st">&quot;pbmc_small&quot;</span>)</span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a><span class="co"># Standard Seurat preprocessing</span></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a>pbmc_small <span class="ot">&lt;-</span> <span class="fu">NormalizeData</span>(pbmc_small)</span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a>pbmc_small <span class="ot">&lt;-</span> <span class="fu">FindVariableFeatures</span>(pbmc_small)</span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a>pbmc_small <span class="ot">&lt;-</span> <span class="fu">ScaleData</span>(pbmc_small)</span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a>pbmc_small <span class="ot">&lt;-</span> <span class="fu">RunPCA</span>(pbmc_small)</span>
<span id="cb9-13"><a href="#cb9-13" tabindex="-1"></a>pbmc_small <span class="ot">&lt;-</span> <span class="fu">FindNeighbors</span>(pbmc_small)</span>
<span id="cb9-14"><a href="#cb9-14" tabindex="-1"></a>pbmc_small <span class="ot">&lt;-</span> <span class="fu">FindClusters</span>(pbmc_small, <span class="at">resolution =</span> <span class="fl">0.5</span>)</span>
<span id="cb9-15"><a href="#cb9-15" tabindex="-1"></a>pbmc_small <span class="ot">&lt;-</span> <span class="fu">RunUMAP</span>(pbmc_small, <span class="at">dims =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>)</span>
<span id="cb9-16"><a href="#cb9-16" tabindex="-1"></a></span>
<span id="cb9-17"><a href="#cb9-17" tabindex="-1"></a><span class="co"># Find markers for each cluster</span></span>
<span id="cb9-18"><a href="#cb9-18" tabindex="-1"></a>pbmc_markers <span class="ot">&lt;-</span> <span class="fu">FindAllMarkers</span>(pbmc_small, <span class="at">only.pos =</span> <span class="cn">TRUE</span>, <span class="at">min.pct =</span> <span class="fl">0.25</span>, <span class="at">logfc.threshold =</span> <span class="fl">0.25</span>)</span>
<span id="cb9-19"><a href="#cb9-19" tabindex="-1"></a></span>
<span id="cb9-20"><a href="#cb9-20" tabindex="-1"></a><span class="co"># Define models to use</span></span>
<span id="cb9-21"><a href="#cb9-21" tabindex="-1"></a>models <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb9-22"><a href="#cb9-22" tabindex="-1"></a>  <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,</span>
<span id="cb9-23"><a href="#cb9-23" tabindex="-1"></a>  <span class="st">&quot;gpt-5&quot;</span>,</span>
<span id="cb9-24"><a href="#cb9-24" tabindex="-1"></a>  <span class="st">&quot;gemini-1.5-pro&quot;</span></span>
<span id="cb9-25"><a href="#cb9-25" tabindex="-1"></a>)</span>
<span id="cb9-26"><a href="#cb9-26" tabindex="-1"></a></span>
<span id="cb9-27"><a href="#cb9-27" tabindex="-1"></a><span class="co"># API keys</span></span>
<span id="cb9-28"><a href="#cb9-28" tabindex="-1"></a>api_keys <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb9-29"><a href="#cb9-29" tabindex="-1"></a>  <span class="at">anthropic =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>),</span>
<span id="cb9-30"><a href="#cb9-30" tabindex="-1"></a>  <span class="at">openai =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;OPENAI_API_KEY&quot;</span>),</span>
<span id="cb9-31"><a href="#cb9-31" tabindex="-1"></a>  <span class="at">gemini =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;GEMINI_API_KEY&quot;</span>)</span>
<span id="cb9-32"><a href="#cb9-32" tabindex="-1"></a>)</span>
<span id="cb9-33"><a href="#cb9-33" tabindex="-1"></a></span>
<span id="cb9-34"><a href="#cb9-34" tabindex="-1"></a><span class="co"># Run annotation with multiple models</span></span>
<span id="cb9-35"><a href="#cb9-35" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb9-36"><a href="#cb9-36" tabindex="-1"></a><span class="cf">for</span> (model <span class="cf">in</span> models) {</span>
<span id="cb9-37"><a href="#cb9-37" tabindex="-1"></a>  provider <span class="ot">&lt;-</span> <span class="fu">get_provider</span>(model)</span>
<span id="cb9-38"><a href="#cb9-38" tabindex="-1"></a>  api_key <span class="ot">&lt;-</span> api_keys[[provider]]</span>
<span id="cb9-39"><a href="#cb9-39" tabindex="-1"></a></span>
<span id="cb9-40"><a href="#cb9-40" tabindex="-1"></a>  results[[model]] <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb9-41"><a href="#cb9-41" tabindex="-1"></a>    <span class="at">input =</span> pbmc_markers,</span>
<span id="cb9-42"><a href="#cb9-42" tabindex="-1"></a>    <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb9-43"><a href="#cb9-43" tabindex="-1"></a>    <span class="at">model =</span> model,</span>
<span id="cb9-44"><a href="#cb9-44" tabindex="-1"></a>    <span class="at">api_key =</span> api_key,</span>
<span id="cb9-45"><a href="#cb9-45" tabindex="-1"></a>    <span class="at">top_gene_count =</span> <span class="dv">10</span></span>
<span id="cb9-46"><a href="#cb9-46" tabindex="-1"></a>  )</span>
<span id="cb9-47"><a href="#cb9-47" tabindex="-1"></a></span>
<span id="cb9-48"><a href="#cb9-48" tabindex="-1"></a>  <span class="co"># Add individual model results to Seurat object</span></span>
<span id="cb9-49"><a href="#cb9-49" tabindex="-1"></a>  column_name <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;cell_type_&quot;</span>, <span class="fu">gsub</span>(<span class="st">&quot;[^a-zA-Z0-9]&quot;</span>, <span class="st">&quot;_&quot;</span>, model))</span>
<span id="cb9-50"><a href="#cb9-50" tabindex="-1"></a>  pbmc_small[[column_name]] <span class="ot">&lt;-</span> plyr<span class="sc">::</span><span class="fu">mapvalues</span>(</span>
<span id="cb9-51"><a href="#cb9-51" tabindex="-1"></a>    <span class="at">x =</span> <span class="fu">as.character</span>(<span class="fu">Idents</span>(pbmc_small)),</span>
<span id="cb9-52"><a href="#cb9-52" tabindex="-1"></a>    <span class="at">from =</span> <span class="fu">names</span>(results[[model]]),</span>
<span id="cb9-53"><a href="#cb9-53" tabindex="-1"></a>    <span class="at">to =</span> results[[model]]</span>
<span id="cb9-54"><a href="#cb9-54" tabindex="-1"></a>  )</span>
<span id="cb9-55"><a href="#cb9-55" tabindex="-1"></a>}</span>
<span id="cb9-56"><a href="#cb9-56" tabindex="-1"></a></span>
<span id="cb9-57"><a href="#cb9-57" tabindex="-1"></a><span class="co"># Create consensus</span></span>
<span id="cb9-58"><a href="#cb9-58" tabindex="-1"></a>consensus_results <span class="ot">&lt;-</span> <span class="fu">interactive_consensus_annotation</span>(</span>
<span id="cb9-59"><a href="#cb9-59" tabindex="-1"></a>  <span class="at">input =</span> pbmc_markers,</span>
<span id="cb9-60"><a href="#cb9-60" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb9-61"><a href="#cb9-61" tabindex="-1"></a>  <span class="at">models =</span> models,  <span class="co"># Use all the models defined above</span></span>
<span id="cb9-62"><a href="#cb9-62" tabindex="-1"></a>  <span class="at">api_keys =</span> api_keys,</span>
<span id="cb9-63"><a href="#cb9-63" tabindex="-1"></a>  <span class="at">controversy_threshold =</span> <span class="fl">0.7</span>,</span>
<span id="cb9-64"><a href="#cb9-64" tabindex="-1"></a>  <span class="at">entropy_threshold =</span> <span class="fl">1.0</span>,</span>
<span id="cb9-65"><a href="#cb9-65" tabindex="-1"></a>  <span class="at">consensus_check_model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span></span>
<span id="cb9-66"><a href="#cb9-66" tabindex="-1"></a>)</span>
<span id="cb9-67"><a href="#cb9-67" tabindex="-1"></a></span>
<span id="cb9-68"><a href="#cb9-68" tabindex="-1"></a><span class="co"># Add consensus results to Seurat object</span></span>
<span id="cb9-69"><a href="#cb9-69" tabindex="-1"></a>pbmc_small<span class="sc">$</span>cell_type_consensus <span class="ot">&lt;-</span> plyr<span class="sc">::</span><span class="fu">mapvalues</span>(</span>
<span id="cb9-70"><a href="#cb9-70" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">as.character</span>(<span class="fu">Idents</span>(pbmc_small)),</span>
<span id="cb9-71"><a href="#cb9-71" tabindex="-1"></a>  <span class="at">from =</span> <span class="fu">names</span>(consensus_results<span class="sc">$</span>final_annotations),</span>
<span id="cb9-72"><a href="#cb9-72" tabindex="-1"></a>  <span class="at">to =</span> consensus_results<span class="sc">$</span>final_annotations</span>
<span id="cb9-73"><a href="#cb9-73" tabindex="-1"></a>)</span>
<span id="cb9-74"><a href="#cb9-74" tabindex="-1"></a></span>
<span id="cb9-75"><a href="#cb9-75" tabindex="-1"></a><span class="co"># Extract consensus metrics from the consensus results</span></span>
<span id="cb9-76"><a href="#cb9-76" tabindex="-1"></a>consensus_metrics <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">names</span>(consensus_results<span class="sc">$</span>initial_results<span class="sc">$</span>consensus_results), <span class="cf">function</span>(cluster_id) {</span>
<span id="cb9-77"><a href="#cb9-77" tabindex="-1"></a>  metrics <span class="ot">&lt;-</span> consensus_results<span class="sc">$</span>initial_results<span class="sc">$</span>consensus_results[[cluster_id]]</span>
<span id="cb9-78"><a href="#cb9-78" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">list</span>(</span>
<span id="cb9-79"><a href="#cb9-79" tabindex="-1"></a>    <span class="at">cluster =</span> cluster_id,</span>
<span id="cb9-80"><a href="#cb9-80" tabindex="-1"></a>    <span class="at">consensus_proportion =</span> metrics<span class="sc">$</span>consensus_proportion,</span>
<span id="cb9-81"><a href="#cb9-81" tabindex="-1"></a>    <span class="at">entropy =</span> metrics<span class="sc">$</span>entropy</span>
<span id="cb9-82"><a href="#cb9-82" tabindex="-1"></a>  ))</span>
<span id="cb9-83"><a href="#cb9-83" tabindex="-1"></a>})</span>
<span id="cb9-84"><a href="#cb9-84" tabindex="-1"></a></span>
<span id="cb9-85"><a href="#cb9-85" tabindex="-1"></a><span class="co"># Convert to data frame for easier handling</span></span>
<span id="cb9-86"><a href="#cb9-86" tabindex="-1"></a>metrics_df <span class="ot">&lt;-</span> <span class="fu">do.call</span>(rbind, <span class="fu">lapply</span>(consensus_metrics, data.frame))</span>
<span id="cb9-87"><a href="#cb9-87" tabindex="-1"></a></span>
<span id="cb9-88"><a href="#cb9-88" tabindex="-1"></a><span class="co"># Add consensus proportion to Seurat object</span></span>
<span id="cb9-89"><a href="#cb9-89" tabindex="-1"></a>pbmc_small<span class="sc">$</span>consensus_proportion <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(plyr<span class="sc">::</span><span class="fu">mapvalues</span>(</span>
<span id="cb9-90"><a href="#cb9-90" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">as.character</span>(<span class="fu">Idents</span>(pbmc_small)),</span>
<span id="cb9-91"><a href="#cb9-91" tabindex="-1"></a>  <span class="at">from =</span> metrics_df<span class="sc">$</span>cluster,</span>
<span id="cb9-92"><a href="#cb9-92" tabindex="-1"></a>  <span class="at">to =</span> metrics_df<span class="sc">$</span>consensus_proportion</span>
<span id="cb9-93"><a href="#cb9-93" tabindex="-1"></a>))</span>
<span id="cb9-94"><a href="#cb9-94" tabindex="-1"></a></span>
<span id="cb9-95"><a href="#cb9-95" tabindex="-1"></a><span class="co"># Add entropy to Seurat object</span></span>
<span id="cb9-96"><a href="#cb9-96" tabindex="-1"></a>pbmc_small<span class="sc">$</span>shannon_entropy <span class="ot">&lt;-</span> <span class="fu">as.numeric</span>(plyr<span class="sc">::</span><span class="fu">mapvalues</span>(</span>
<span id="cb9-97"><a href="#cb9-97" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">as.character</span>(<span class="fu">Idents</span>(pbmc_small)),</span>
<span id="cb9-98"><a href="#cb9-98" tabindex="-1"></a>  <span class="at">from =</span> metrics_df<span class="sc">$</span>cluster,</span>
<span id="cb9-99"><a href="#cb9-99" tabindex="-1"></a>  <span class="at">to =</span> metrics_df<span class="sc">$</span>entropy</span>
<span id="cb9-100"><a href="#cb9-100" tabindex="-1"></a>))</span>
<span id="cb9-101"><a href="#cb9-101" tabindex="-1"></a></span>
<span id="cb9-102"><a href="#cb9-102" tabindex="-1"></a><span class="co"># Visualize results</span></span>
<span id="cb9-103"><a href="#cb9-103" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">DimPlot</span>(pbmc_small, <span class="at">group.by =</span> <span class="st">&quot;cell_type_consensus&quot;</span>, <span class="at">label =</span> <span class="cn">TRUE</span>, <span class="at">repel =</span> <span class="cn">TRUE</span>) <span class="sc">+</span></span>
<span id="cb9-104"><a href="#cb9-104" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Cell Type Annotations&quot;</span>) <span class="sc">+</span></span>
<span id="cb9-105"><a href="#cb9-105" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>))</span>
<span id="cb9-106"><a href="#cb9-106" tabindex="-1"></a></span>
<span id="cb9-107"><a href="#cb9-107" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">FeaturePlot</span>(pbmc_small, <span class="at">features =</span> <span class="st">&quot;consensus_proportion&quot;</span>, <span class="at">cols =</span> <span class="fu">c</span>(<span class="st">&quot;yellow&quot;</span>, <span class="st">&quot;green&quot;</span>, <span class="st">&quot;blue&quot;</span>)) <span class="sc">+</span></span>
<span id="cb9-108"><a href="#cb9-108" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Consensus Proportion&quot;</span>) <span class="sc">+</span></span>
<span id="cb9-109"><a href="#cb9-109" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>))</span>
<span id="cb9-110"><a href="#cb9-110" tabindex="-1"></a></span>
<span id="cb9-111"><a href="#cb9-111" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">FeaturePlot</span>(pbmc_small, <span class="at">features =</span> <span class="st">&quot;shannon_entropy&quot;</span>, <span class="at">cols =</span> <span class="fu">c</span>(<span class="st">&quot;red&quot;</span>, <span class="st">&quot;orange&quot;</span>)) <span class="sc">+</span></span>
<span id="cb9-112"><a href="#cb9-112" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">&quot;Shannon Entropy&quot;</span>) <span class="sc">+</span></span>
<span id="cb9-113"><a href="#cb9-113" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>))</span>
<span id="cb9-114"><a href="#cb9-114" tabindex="-1"></a></span>
<span id="cb9-115"><a href="#cb9-115" tabindex="-1"></a><span class="co"># Combine plots</span></span>
<span id="cb9-116"><a href="#cb9-116" tabindex="-1"></a>p1 <span class="sc">|</span> p2 <span class="sc">|</span> p3</span></code></pre></div>
</div>
<div id="advanced-parameter-tuning" class="section level2">
<h2>Advanced Parameter Tuning</h2>
<div id="adjusting-top_gene_count" class="section level3">
<h3>Adjusting top_gene_count</h3>
<p>The <code>top_gene_count</code> parameter controls how many top
marker genes per cluster are used for annotation:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="co"># Using more genes (better for well-characterized tissues)</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>results_more_genes <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>  <span class="at">input =</span> pbmc_markers,</span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,</span>
<span id="cb10-6"><a href="#cb10-6" tabindex="-1"></a>  <span class="at">api_key =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>),</span>
<span id="cb10-7"><a href="#cb10-7" tabindex="-1"></a>  <span class="at">top_gene_count =</span> <span class="dv">20</span>  <span class="co"># Using more genes</span></span>
<span id="cb10-8"><a href="#cb10-8" tabindex="-1"></a>)</span>
<span id="cb10-9"><a href="#cb10-9" tabindex="-1"></a></span>
<span id="cb10-10"><a href="#cb10-10" tabindex="-1"></a><span class="co"># Using fewer genes (better for noisy data)</span></span>
<span id="cb10-11"><a href="#cb10-11" tabindex="-1"></a>results_fewer_genes <span class="ot">&lt;-</span> <span class="fu">annotate_cell_types</span>(</span>
<span id="cb10-12"><a href="#cb10-12" tabindex="-1"></a>  <span class="at">input =</span> pbmc_markers,</span>
<span id="cb10-13"><a href="#cb10-13" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb10-14"><a href="#cb10-14" tabindex="-1"></a>  <span class="at">model =</span> <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>,</span>
<span id="cb10-15"><a href="#cb10-15" tabindex="-1"></a>  <span class="at">api_key =</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>),</span>
<span id="cb10-16"><a href="#cb10-16" tabindex="-1"></a>  <span class="at">top_gene_count =</span> <span class="dv">5</span>   <span class="co"># Using fewer genes</span></span>
<span id="cb10-17"><a href="#cb10-17" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="adjusting-controversy_threshold" class="section level3">
<h3>Adjusting controversy_threshold</h3>
<p>The <code>controversy_threshold</code> parameter in the
<code>interactive_consensus_annotation</code> function controls which
clusters are considered controversial and require discussion:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="co"># Example of using interactive_consensus_annotation with different controversy thresholds</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="co"># Lower threshold (more clusters will be discussed)</span></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a>consensus_results_low_threshold <span class="ot">&lt;-</span> <span class="fu">interactive_consensus_annotation</span>(</span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a>  <span class="at">input =</span> pbmc_markers,</span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a>  <span class="at">models =</span> <span class="fu">c</span>(<span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>, <span class="st">&quot;gpt-5&quot;</span>, <span class="st">&quot;gemini-2.0-flash&quot;</span>),</span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a>  <span class="at">api_keys =</span> <span class="fu">list</span>(</span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a>    <span class="st">&quot;anthropic&quot;</span> <span class="ot">=</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>),</span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a>    <span class="st">&quot;openai&quot;</span> <span class="ot">=</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;OPENAI_API_KEY&quot;</span>),</span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a>    <span class="st">&quot;gemini&quot;</span> <span class="ot">=</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;GEMINI_API_KEY&quot;</span>)</span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a>  ),</span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a>  <span class="at">controversy_threshold =</span> <span class="fl">0.3</span>  <span class="co"># Lower threshold - more clusters will be discussed</span></span>
<span id="cb11-13"><a href="#cb11-13" tabindex="-1"></a>)</span>
<span id="cb11-14"><a href="#cb11-14" tabindex="-1"></a></span>
<span id="cb11-15"><a href="#cb11-15" tabindex="-1"></a><span class="co"># Higher threshold (fewer clusters will be discussed)</span></span>
<span id="cb11-16"><a href="#cb11-16" tabindex="-1"></a>consensus_results_high_threshold <span class="ot">&lt;-</span> <span class="fu">interactive_consensus_annotation</span>(</span>
<span id="cb11-17"><a href="#cb11-17" tabindex="-1"></a>  <span class="at">input =</span> pbmc_markers,</span>
<span id="cb11-18"><a href="#cb11-18" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;human PBMC&quot;</span>,</span>
<span id="cb11-19"><a href="#cb11-19" tabindex="-1"></a>  <span class="at">models =</span> <span class="fu">c</span>(<span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>, <span class="st">&quot;gpt-5&quot;</span>, <span class="st">&quot;gemini-2.0-flash&quot;</span>),</span>
<span id="cb11-20"><a href="#cb11-20" tabindex="-1"></a>  <span class="at">api_keys =</span> <span class="fu">list</span>(</span>
<span id="cb11-21"><a href="#cb11-21" tabindex="-1"></a>    <span class="st">&quot;anthropic&quot;</span> <span class="ot">=</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;ANTHROPIC_API_KEY&quot;</span>),</span>
<span id="cb11-22"><a href="#cb11-22" tabindex="-1"></a>    <span class="st">&quot;openai&quot;</span> <span class="ot">=</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;OPENAI_API_KEY&quot;</span>),</span>
<span id="cb11-23"><a href="#cb11-23" tabindex="-1"></a>    <span class="st">&quot;gemini&quot;</span> <span class="ot">=</span> <span class="fu">Sys.getenv</span>(<span class="st">&quot;GEMINI_API_KEY&quot;</span>)</span>
<span id="cb11-24"><a href="#cb11-24" tabindex="-1"></a>  ),</span>
<span id="cb11-25"><a href="#cb11-25" tabindex="-1"></a>  <span class="at">controversy_threshold =</span> <span class="fl">0.7</span>  <span class="co"># Higher threshold - fewer clusters will be discussed</span></span>
<span id="cb11-26"><a href="#cb11-26" tabindex="-1"></a>)</span></code></pre></div>
</div>
</div>
<div id="performance-considerations" class="section level2">
<h2>Performance Considerations</h2>
<div id="api-rate-limits-and-costs" class="section level3">
<h3>API Rate Limits and Costs</h3>
<p>Different LLM providers have different rate limits and pricing:</p>
<ul>
<li><strong>OpenAI</strong>: Higher rate limits but can be more
expensive</li>
<li><strong>Anthropic</strong>: Good balance of rate limits and
cost</li>
<li><strong>Google</strong>: Competitive pricing with good rate
limits</li>
<li><strong>Others</strong>: Generally lower cost but may have stricter
rate limits</li>
</ul>
<p>To manage costs and rate limits:</p>
<ol style="list-style-type: decimal">
<li>Use caching to avoid redundant API calls</li>
<li>Start with a single model for exploration</li>
<li>Use more economical models for initial testing</li>
<li>Reserve multi-model consensus for final analysis</li>
</ol>
</div>
<div id="execution-time" class="section level3">
<h3>Execution Time</h3>
<p>Typical execution times:</p>
<ul>
<li>Single model annotation: 5-30 seconds per cluster</li>
<li>Multi-model consensus: 1-5 minutes for a typical dataset</li>
<li>Discussion process: Additional 1-3 minutes per controversial
cluster</li>
</ul>
<p>To improve performance:</p>
<ol style="list-style-type: decimal">
<li>Use a smaller <code>top_gene_count</code> for faster execution</li>
<li>Enable caching to reuse results</li>
<li>Use a higher <code>controversy_threshold</code> to reduce the number
of clusters that require discussion</li>
</ol>
</div>
</div>
<div id="troubleshooting" class="section level2">
<h2>Troubleshooting</h2>
<div id="common-issues-with-openrouter" class="section level3">
<h3>Common Issues with OpenRouter</h3>
<p>If you encounter “No auth credentials found” errors with
OpenRouter:</p>
<ul>
<li>Verify your API key is correct and active</li>
<li>Ensure you’re using the correct format for the model name (e.g.,
<code>provider/model-name:free</code>)</li>
<li>Try a different free model, as some models may have access
restrictions</li>
</ul>
</div>
<div id="non-english-model-error-messages" class="section level3">
<h3>Non-English Model Error Messages</h3>
<p>If you see “non-character parameter” errors:</p>
<ul>
<li>This typically occurs when using non-English language models like
<code>thudm/glm-z1-9b:free</code> for consensus checking</li>
<li>Switch to an English-optimized model like
<code>meta-llama/llama-4-maverick:free</code> for consensus
checking</li>
</ul>
</div>
<div id="empty-results" class="section level3">
<h3>Empty Results</h3>
<p>If a model returns empty results:</p>
<ul>
<li>Try increasing the timeout or retry with the same model</li>
<li>Some models like <code>deepseek/deepseek-r1:free</code> may
occasionally return empty results</li>
<li>Switch to a more reliable model if the problem persists</li>
</ul>
</div>
</div>
<div id="next-steps" class="section level2">
<h2>Next Steps</h2>
<p>Now that you understand the detailed usage of mLLMCelltype, you can
explore:</p>
<ul>
<li><a href="https://cafferyang.com/mLLMCelltype/articles/consensus-principles.html">Consensus
Annotation Principles</a>: Learn about the technical principles</li>
<li><a href="https://cafferyang.com/mLLMCelltype/articles/visualization-guide.html">Visualization
Guide</a>: Create publication-ready visualizations</li>
<li><a href="https://cafferyang.com/mLLMCelltype/articles/faq.html">FAQ</a>:
Find answers to common questions</li>
<li><a href="https://cafferyang.com/mLLMCelltype/articles/advanced-features.html">Advanced
Features</a>: Explore hierarchical annotation and other advanced
features</li>
</ul>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
