<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Why Choose Consensus? The Scientific Foundation of Multi-LLM Annotation</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Why Choose Consensus? The Scientific
Foundation of Multi-LLM Annotation</h1>



<div id="why-choose-consensus-the-scientific-foundation-of-multi-llm-annotation" class="section level1">
<h1>Why Choose Consensus? The Scientific Foundation of Multi-LLM
Annotation</h1>
<p>Multi-LLM consensus can improve annotation accuracy by combining the
strengths of diverse AI models while reducing the impact of individual
model limitations (see Yang et al., 2025).</p>
<div id="the-challenge-with-single-model-approaches" class="section level2">
<h2>The Challenge with Single-Model Approaches</h2>
<p>Traditional single-model annotation systems face inherent
limitations:</p>
<div id="accuracy-limitations" class="section level3">
<h3>Accuracy Limitations</h3>
<ul>
<li><strong>Single-point failure</strong>: One model’s bias affects all
results</li>
<li><strong>Limited perspective</strong>: Each model has unique
strengths and blind spots</li>
<li><strong>Inconsistent performance</strong>: Varies across cell types
and tissues</li>
</ul>
</div>
<div id="reliability-issues" class="section level3">
<h3>Reliability Issues</h3>
<ul>
<li><strong>Model hallucinations</strong>: Confident but incorrect
predictions</li>
<li><strong>Lack of uncertainty</strong>: Difficult to identify
questionable annotations</li>
<li><strong>Reproducibility challenges</strong>: Different model
versions may yield different results</li>
</ul>
</div>
</div>
<div id="the-consensus-approach-inspired-by-scientific-peer-review" class="section level2">
<h2>The Consensus Approach: Inspired by Scientific Peer Review</h2>
<p>mLLMCelltype’s consensus framework is analogous to the peer review
process in scientific publishing.</p>
<div id="the-scientific-parallel" class="section level3">
<h3>The Scientific Parallel</h3>
<p>Just as scientific papers benefit from multiple expert reviewers,
cell annotations can benefit from multiple AI models:</p>
<table>
<thead>
<tr class="header">
<th>Scientific Peer Review</th>
<th>mLLMCelltype Consensus</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Multiple expert reviewers</td>
<td>Multiple LLM models</td>
</tr>
<tr class="even">
<td>Diverse perspectives</td>
<td>Different training approaches</td>
</tr>
<tr class="odd">
<td>Debate and discussion</td>
<td>Structured deliberation</td>
</tr>
<tr class="even">
<td>Consensus building</td>
<td>Agreement quantification</td>
</tr>
<tr class="odd">
<td>Quality assurance</td>
<td>Uncertainty metrics</td>
</tr>
</tbody>
</table>
</div>
<div id="how-it-works" class="section level3">
<h3>How It Works</h3>
<p><strong>1. Error Detection Through Cross-Validation</strong> - Models
check each other’s work - Individual model biases can be averaged out -
Outlier predictions are identified</p>
<p><strong>2. Transparent Uncertainty Quantification</strong> -
<strong>Consensus Proportion (CP)</strong>: Measures inter-model
agreement - <strong>Shannon Entropy</strong>: Quantifies prediction
uncertainty - <strong>Controversy Detection</strong>: Automatically
identifies clusters requiring expert review</p>
</div>
</div>
<div id="why-multiple-perspectives-help" class="section level2">
<h2>Why Multiple Perspectives Help</h2>
<p>Cell type annotation involves:</p>
<ul>
<li><strong>Marker gene interpretation</strong>: Different models may
have different strengths across gene families</li>
<li><strong>Context understanding</strong>: Various models may capture
different biological contexts</li>
<li><strong>Rare cell types</strong>: Ensemble approaches can improve
detection of uncommon populations</li>
<li><strong>Batch effects</strong>: Multiple models may provide
robustness against technical artifacts</li>
</ul>
<p>For benchmark results, see Yang et al. (2025):</p>
<p>Yang, C., Zhang, X., &amp; Chen, J. (2025). Large Language Model
Consensus Substantially Improves the Cell Type Annotation Accuracy for
scRNA-seq Data. <em>bioRxiv</em>. <a href="https://doi.org/10.1101/2025.04.10.647852" class="uri">https://doi.org/10.1101/2025.04.10.647852</a></p>
</div>
<div id="cost-considerations" class="section level2">
<h2>Cost Considerations</h2>
<p>The two-stage approach can reduce API calls when models agree
early:</p>
<ul>
<li><strong>Stage 1</strong>: Initial consensus check – clusters where
models agree skip further processing</li>
<li><strong>Stage 2</strong>: Deliberation only for clusters without
initial agreement</li>
<li><strong>Caching</strong>: Results can be reused across similar
analyses</li>
</ul>
<p>This means the cost overhead of using multiple models is partially
offset by skipping deliberation for clear cases.</p>
</div>
<div id="technical-implementation" class="section level2">
<h2>Technical Implementation</h2>
<div id="the-three-stage-process" class="section level3">
<h3>The Three-Stage Process</h3>
<p><strong>Stage 1: Independent Analysis</strong> Each LLM analyzes
marker genes and provides: - Cell type predictions - Confidence scores -
Reasoning chains</p>
<p><strong>Stage 2: Consensus Building</strong> The system: - Compares
predictions across models - Identifies areas of agreement and
disagreement - Calculates uncertainty metrics</p>
<p><strong>Stage 3: Deliberation (when needed)</strong> For
controversial clusters: - Models share their reasoning - Structured
debate occurs - Final consensus emerges</p>
</div>
<div id="quality-metrics" class="section level3">
<h3>Quality Metrics</h3>
<ul>
<li><strong>Semantic similarity analysis</strong>: Ensures meaningful
disagreements are detected</li>
<li><strong>Evidence-based reasoning</strong>: All predictions include
supporting evidence</li>
<li><strong>Iterative refinement</strong>: Multiple rounds of discussion
when needed</li>
</ul>
</div>
</div>
<div id="when-to-choose-consensus" class="section level2">
<h2>When to Choose Consensus</h2>
<p><strong>Consensus may be preferable when:</strong> - Uncertainty
quantification is needed - Datasets involve novel or complex tissues -
Results will be published or used in downstream analyses - Identifying
low-confidence annotations is important</p>
<p><strong>Consider alternatives when:</strong> - Quick exploratory
analysis is the goal - Datasets are well-characterized with clear
markers - API budget is very limited - Proof-of-concept work in early
stages</p>
</div>
<div id="quick-start-example" class="section level2">
<h2>Quick Start Example</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(mLLMCelltype)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="co"># Load your single-cell data</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a>results <span class="ot">&lt;-</span> <span class="fu">interactive_consensus_annotation</span>(</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a>  <span class="at">seurat_obj =</span> your_data,</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a>  <span class="at">tissue_name =</span> <span class="st">&quot;PBMC&quot;</span>,</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a>  <span class="at">models =</span> <span class="fu">c</span>(<span class="st">&quot;gpt-4o&quot;</span>, <span class="st">&quot;claude-sonnet-4-5-20250929&quot;</span>, <span class="st">&quot;gemini-2.5-pro&quot;</span>),</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a>  <span class="at">consensus_method =</span> <span class="st">&quot;iterative&quot;</span></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a>)</span></code></pre></div>
<div id="understanding-your-results" class="section level3">
<h3>Understanding Your Results</h3>
<ul>
<li><strong>High consensus (CP &gt; 0.8)</strong>: Reliable
annotations</li>
<li><strong>Medium consensus (0.5 &lt; CP &lt; 0.8)</strong>: Review
recommended</li>
<li><strong>Low consensus (CP &lt; 0.5)</strong>: Expert validation
needed</li>
</ul>
</div>
</div>
<div id="summary" class="section level2">
<h2>Summary</h2>
<p>The consensus approach provides a framework for combining multiple
LLM predictions with built-in uncertainty quantification. As new models
become available, the framework can incorporate them without changes to
the overall methodology.</p>
</div>
<div id="learn-more" class="section level2">
<h2>Learn More</h2>
<ul>
<li><a href="https://cafferyang.com/mLLMCelltype/articles/getting-started.html">Getting
Started Guide</a></li>
<li><a href="https://cafferyang.com/mLLMCelltype/articles/vs-single-agent.html">Consensus
vs Single-Agent Methods</a></li>
<li><a href="https://cafferyang.com/mLLMCelltype/articles/advanced-features.html">Performance
Benchmarks</a></li>
<li><a href="https://cafferyang.com/mLLMCelltype/reference/index.html">API
Reference</a></li>
</ul>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
